{
 "cells": [
  {
   "cell_type": "markdown",
   "source": [
    "<h1>Table of Contents<span class=\"tocSkip\"></span></h1>\n",
    "<div class=\"toc\"><ul class=\"toc-item\"><li><span><a href=\"#Read-pathbank-data\" data-toc-modified-id=\"Read-pathbank-data-1\"><span class=\"toc-item-num\">1&nbsp;&nbsp;</span>Read pathbank data</a></span><ul class=\"toc-item\"><li><span><a href=\"#Read-Pathbank-Pathway-metadata\" data-toc-modified-id=\"Read-Pathbank-Pathway-metadata-1.1\"><span class=\"toc-item-num\">1.1&nbsp;&nbsp;</span>Read Pathbank Pathway metadata</a></span></li><li><span><a href=\"#Generate-species-pathway-ID\" data-toc-modified-id=\"Generate-species-pathway-ID-1.2\"><span class=\"toc-item-num\">1.2&nbsp;&nbsp;</span>Generate species pathway ID</a></span></li><li><span><a href=\"#Generate-human-pathway-graph\" data-toc-modified-id=\"Generate-human-pathway-graph-1.3\"><span class=\"toc-item-num\">1.3&nbsp;&nbsp;</span>Generate human pathway graph</a></span><ul class=\"toc-item\"><li><span><a href=\"#Read-human-pathway\" data-toc-modified-id=\"Read-human-pathway-1.3.1\"><span class=\"toc-item-num\">1.3.1&nbsp;&nbsp;</span>Read human pathway</a></span></li><li><span><a href=\"#Analysis-of-protein-classes-and-metabolite-types\" data-toc-modified-id=\"Analysis-of-protein-classes-and-metabolite-types-1.3.2\"><span class=\"toc-item-num\">1.3.2&nbsp;&nbsp;</span>Analysis of protein classes and metabolite types</a></span></li></ul></li></ul></li></ul></div>"
   ],
   "metadata": {
    "toc": true
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "**Run the notebook in conda tf env** <br>\n",
    "**Conda activate tf**"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "source": [
    "import pandas as pd\n",
    "\n",
    "import numpy as np\n",
    "import networkx as nx\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as mpatches\n",
    "import scipy.sparse as sp\n",
    "import csv\n",
    "import seaborn as sns\n",
    "sns.set_style('whitegrid')\n",
    "import glob\n",
    "from rdkit import Chem\n",
    "from rdkit.Chem import MACCSkeys\n",
    "import json\n",
    "from pathlib import Path\n",
    "import pickle\n",
    "from rdkit.Chem import AllChem\n",
    "from rdkit.Chem import DataStructs"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "RDKit WARNING: [19:12:49] Enabling RDKit 2019.09.3 jupyter extensions\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "source": [
    "import pandas as pd\n",
    "import pickle\n",
    "import networkx as nx\n",
    "import glob\n",
    "from rdkit import Chem\n",
    "from rdkit.Chem import MACCSkeys\n",
    "from rdkit.Chem import AllChem\n",
    "from rdkit.Chem import DataStructs\n",
    "import numpy as np\n",
    "from sklearn.decomposition import PCA\n",
    "from Bio import SeqIO\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "# Generation of MPI network"
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Read Pathbank pathway metadata"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "source": [
    "meta_data_dir = '../Training_data/pathbank/Pathbank_Meta_Data/'\n",
    "meta_data = pd.read_csv(meta_data_dir+'pathbank_pathways.csv')\n",
    "meta_data.head(2)"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SMPDB ID</th>\n",
       "      <th>PW ID</th>\n",
       "      <th>Name</th>\n",
       "      <th>Subject</th>\n",
       "      <th>Description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SMP0000055</td>\n",
       "      <td>PW000001</td>\n",
       "      <td>Alanine Metabolism</td>\n",
       "      <td>Metabolic</td>\n",
       "      <td>Alanine (L-Alanine) is an α-amino acid that is...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SMP0000067</td>\n",
       "      <td>PW000002</td>\n",
       "      <td>Aspartate Metabolism</td>\n",
       "      <td>Metabolic</td>\n",
       "      <td>Aspartate is synthesized by transamination of ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     SMPDB ID     PW ID                  Name    Subject  \\\n",
       "0  SMP0000055  PW000001    Alanine Metabolism  Metabolic   \n",
       "1  SMP0000067  PW000002  Aspartate Metabolism  Metabolic   \n",
       "\n",
       "                                         Description  \n",
       "0  Alanine (L-Alanine) is an α-amino acid that is...  \n",
       "1  Aspartate is synthesized by transamination of ...  "
      ]
     },
     "metadata": {},
     "execution_count": 3
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Read Pathway protein/metabolite data"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "source": [
    "metabolite_data = pd.read_csv(meta_data_dir+'pathbank_all_metabolites.csv') \n",
    "protein_data = pd.read_csv(meta_data_dir+'pathbank_all_proteins.csv')\n",
    "main_metabolite_data = pd.read_csv(meta_data_dir+'pathbank_primary_metabolites.csv') \n",
    "main_protein_data = pd.read_csv(meta_data_dir+'pathbank_primary_proteins.csv')"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "/Users/chengwang/opt/anaconda3/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3165: DtypeWarning: Columns (6,7,8,10) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "source": [
    "metabolite_data.head(2)"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PathBank ID</th>\n",
       "      <th>Pathway Name</th>\n",
       "      <th>Pathway Subject</th>\n",
       "      <th>Species</th>\n",
       "      <th>Metabolite ID</th>\n",
       "      <th>Metabolite Name</th>\n",
       "      <th>HMDB ID</th>\n",
       "      <th>KEGG ID</th>\n",
       "      <th>ChEBI ID</th>\n",
       "      <th>DrugBank ID</th>\n",
       "      <th>CAS</th>\n",
       "      <th>Formula</th>\n",
       "      <th>IUPAC</th>\n",
       "      <th>SMILES</th>\n",
       "      <th>InChI</th>\n",
       "      <th>InChI Key</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SMP0000055</td>\n",
       "      <td>Alanine Metabolism</td>\n",
       "      <td>Metabolic</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>PW_C000414</td>\n",
       "      <td>Adenosine triphosphate</td>\n",
       "      <td>HMDB0000538</td>\n",
       "      <td>C00002</td>\n",
       "      <td>15422.0</td>\n",
       "      <td>DB00171</td>\n",
       "      <td>56-65-5</td>\n",
       "      <td>C10H16N5O13P3</td>\n",
       "      <td>({[({[(2R,3S,4R,5R)-5-(6-amino-9H-purin-9-yl)-...</td>\n",
       "      <td>NC1=NC=NC2=C1N=CN2[C@@H]1O[C@H](COP(O)(=O)OP(O...</td>\n",
       "      <td>InChI=1S/C10H16N5O13P3/c11-8-5-9(13-2-12-8)15(...</td>\n",
       "      <td>ZKHQWZAMYRWXGA-KQYNXXCUSA-N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SMP0000055</td>\n",
       "      <td>Alanine Metabolism</td>\n",
       "      <td>Metabolic</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>PW_C000105</td>\n",
       "      <td>L-Alanine</td>\n",
       "      <td>HMDB0000161</td>\n",
       "      <td>C00041</td>\n",
       "      <td>16977.0</td>\n",
       "      <td>DB00160</td>\n",
       "      <td>56-41-7</td>\n",
       "      <td>C3H7NO2</td>\n",
       "      <td>(2S)-2-aminopropanoic acid</td>\n",
       "      <td>C[C@H](N)C(O)=O</td>\n",
       "      <td>InChI=1S/C3H7NO2/c1-2(4)3(5)6/h2H,4H2,1H3,(H,5...</td>\n",
       "      <td>QNAYBMKLOCPYGJ-REOHCLBHSA-N</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  PathBank ID        Pathway Name Pathway Subject       Species Metabolite ID  \\\n",
       "0  SMP0000055  Alanine Metabolism       Metabolic  Homo sapiens    PW_C000414   \n",
       "1  SMP0000055  Alanine Metabolism       Metabolic  Homo sapiens    PW_C000105   \n",
       "\n",
       "          Metabolite Name      HMDB ID KEGG ID  ChEBI ID DrugBank ID      CAS  \\\n",
       "0  Adenosine triphosphate  HMDB0000538  C00002   15422.0     DB00171  56-65-5   \n",
       "1               L-Alanine  HMDB0000161  C00041   16977.0     DB00160  56-41-7   \n",
       "\n",
       "         Formula                                              IUPAC  \\\n",
       "0  C10H16N5O13P3  ({[({[(2R,3S,4R,5R)-5-(6-amino-9H-purin-9-yl)-...   \n",
       "1        C3H7NO2                         (2S)-2-aminopropanoic acid   \n",
       "\n",
       "                                              SMILES  \\\n",
       "0  NC1=NC=NC2=C1N=CN2[C@@H]1O[C@H](COP(O)(=O)OP(O...   \n",
       "1                                    C[C@H](N)C(O)=O   \n",
       "\n",
       "                                               InChI  \\\n",
       "0  InChI=1S/C10H16N5O13P3/c11-8-5-9(13-2-12-8)15(...   \n",
       "1  InChI=1S/C3H7NO2/c1-2(4)3(5)6/h2H,4H2,1H3,(H,5...   \n",
       "\n",
       "                     InChI Key  \n",
       "0  ZKHQWZAMYRWXGA-KQYNXXCUSA-N  \n",
       "1  QNAYBMKLOCPYGJ-REOHCLBHSA-N  "
      ]
     },
     "metadata": {},
     "execution_count": 5
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "source": [
    "protein_data.head(2)"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PathBank ID</th>\n",
       "      <th>Pathway Name</th>\n",
       "      <th>Pathway Subject</th>\n",
       "      <th>Species</th>\n",
       "      <th>Uniprot ID</th>\n",
       "      <th>Protein Name</th>\n",
       "      <th>HMDBP ID</th>\n",
       "      <th>DrugBank ID</th>\n",
       "      <th>GenBank ID</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Locus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SMP0000055</td>\n",
       "      <td>Alanine Metabolism</td>\n",
       "      <td>Metabolic</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>P49588</td>\n",
       "      <td>Alanine--tRNA ligase, cytoplasmic</td>\n",
       "      <td>HMDBP00625</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AC012184</td>\n",
       "      <td>AARS</td>\n",
       "      <td>16q22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SMP0000055</td>\n",
       "      <td>Alanine Metabolism</td>\n",
       "      <td>Metabolic</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>P24298</td>\n",
       "      <td>Alanine aminotransferase 1</td>\n",
       "      <td>HMDBP00850</td>\n",
       "      <td>NaN</td>\n",
       "      <td>U70732</td>\n",
       "      <td>GPT</td>\n",
       "      <td>8q24.3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  PathBank ID        Pathway Name Pathway Subject       Species Uniprot ID  \\\n",
       "0  SMP0000055  Alanine Metabolism       Metabolic  Homo sapiens     P49588   \n",
       "1  SMP0000055  Alanine Metabolism       Metabolic  Homo sapiens     P24298   \n",
       "\n",
       "                        Protein Name    HMDBP ID DrugBank ID GenBank ID  \\\n",
       "0  Alanine--tRNA ligase, cytoplasmic  HMDBP00625         NaN   AC012184   \n",
       "1         Alanine aminotransferase 1  HMDBP00850         NaN     U70732   \n",
       "\n",
       "  Gene Name   Locus  \n",
       "0      AARS   16q22  \n",
       "1       GPT  8q24.3  "
      ]
     },
     "metadata": {},
     "execution_count": 6
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "source": [
    "for i in protein_data['Gene Name'].tolist()[:10]:\n",
    "    print(i)"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "AARS\n",
      "GPT\n",
      "PC\n",
      "AGXT\n",
      "AARS2\n",
      "MPC1\n",
      "ABAT\n",
      "GAD1\n",
      "ASNS\n",
      "NARS\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "source": [
    "import requests"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Obtain features of metabolites and proteins"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "source": [
    "mets_dict = dict(zip(metabolite_data['Metabolite ID'].tolist(),metabolite_data['SMILES'].tolist()))\n",
    "len(mets_dict.keys())\n",
    "mets_vec = {}\n",
    "mets_proc = []\n",
    "mets_fp = []\n",
    "c = 0\n",
    "for k,v in mets_dict.items():\n",
    "    smile = v\n",
    "    try:\n",
    "        mol = Chem.MolFromSmiles(smile)\n",
    "        counts = mol.GetNumAtoms()\n",
    "        fp = AllChem.GetMorganFingerprintAsBitVect(mol, 2, nBits=1024)\n",
    "        array = np.zeros((0, ), dtype=np.int8)\n",
    "        DataStructs.ConvertToNumpyArray(fp, array)\n",
    "        mets_vec[k] = (array,counts)\n",
    "        mets_proc.append(k)\n",
    "        mets_fp.append(array)\n",
    "        c += 1\n",
    "    except:\n",
    "        mets_vec[k] = None\n",
    "        c += 1\n",
    "mets_fp = np.asarray(mets_fp)\n",
    "mets_fp.shape"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "RDKit WARNING: [19:13:03] WARNING: not removing hydrogen atom without neighbors\n",
      "RDKit ERROR: [19:13:03] Explicit valence for atom # 26 N, 4, is greater than permitted\n",
      "RDKit ERROR: [19:13:03] Explicit valence for atom # 1 N, 4, is greater than permitted\n",
      "RDKit ERROR: [19:13:03] Explicit valence for atom # 8 N, 4, is greater than permitted\n",
      "RDKit ERROR: [19:13:03] Explicit valence for atom # 28 N, 4, is greater than permitted\n",
      "RDKit ERROR: [19:13:03] Explicit valence for atom # 31 O, 3, is greater than permitted\n",
      "RDKit ERROR: [19:13:03] Explicit valence for atom # 4 N, 4, is greater than permitted\n",
      "RDKit ERROR: [19:13:04] Explicit valence for atom # 21 N, 4, is greater than permitted\n",
      "RDKit ERROR: [19:13:05] Explicit valence for atom # 17 O, 3, is greater than permitted\n"
     ]
    },
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "(78261, 1024)"
      ]
     },
     "metadata": {},
     "execution_count": 9
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "source": [
    "# Get PCA transformed data of metabolites\n",
    "pca = PCA(n_components=1024)\n",
    "pca.fit(mets_fp)\n",
    "pca_mets_fp = pca.transform(mets_fp)\n",
    "pca_mets_vec_1024 = {}\n",
    "for i in range(len(mets_proc)):\n",
    "    k = mets_proc[i]\n",
    "    pca_mets_vec_1024[k] = pca_mets_fp[i]\n",
    "\n",
    "pca = PCA(n_components=512)\n",
    "pca.fit(mets_fp)\n",
    "pca_mets_fp = pca.transform(mets_fp)\n",
    "pca_mets_vec_512 = {}\n",
    "for i in range(len(mets_proc)):\n",
    "    k = mets_proc[i]\n",
    "    pca_mets_vec_512[k] = pca_mets_fp[i]\n",
    "\n",
    "\n",
    "pca = PCA(n_components=256)\n",
    "pca.fit(mets_fp)\n",
    "pca_mets_fp = pca.transform(mets_fp)\n",
    "pca_mets_vec_256 = {}\n",
    "for i in range(len(mets_proc)):\n",
    "    k = mets_proc[i]\n",
    "    pca_mets_vec_256[k] = pca_mets_fp[i]\n",
    "\n",
    "pca = PCA(n_components=128)\n",
    "pca.fit(mets_fp)\n",
    "pca_mets_fp = pca.transform(mets_fp)\n",
    "pca_mets_vec_128 = {}\n",
    "for i in range(len(mets_proc)):\n",
    "    k = mets_proc[i]\n",
    "    pca_mets_vec_128[k] = pca_mets_fp[i]\n",
    "\n",
    "pca = PCA(n_components=64)\n",
    "pca.fit(mets_fp)\n",
    "pca_mets_fp = pca.transform(mets_fp)\n",
    "pca_mets_vec_64 = {}\n",
    "for i in range(len(mets_proc)):\n",
    "    k = mets_proc[i]\n",
    "    pca_mets_vec_64[k] = pca_mets_fp[i]\n",
    "\n",
    "pca = PCA(n_components=32)\n",
    "pca.fit(mets_fp)\n",
    "pca_mets_fp = pca.transform(mets_fp)\n",
    "pca_mets_vec_32 = {}\n",
    "for i in range(len(mets_proc)):\n",
    "    k = mets_proc[i]\n",
    "    pca_mets_vec_32[k] = pca_mets_fp[i]\n",
    "\n",
    "pca = PCA(n_components=16)\n",
    "pca.fit(mets_fp)\n",
    "pca_mets_fp = pca.transform(mets_fp)\n",
    "pca_mets_vec_16 = {}\n",
    "for i in range(len(mets_proc)):\n",
    "    k = mets_proc[i]\n",
    "    pca_mets_vec_16[k] = pca_mets_fp[i]\n",
    "\n",
    "pca = PCA(n_components=2)\n",
    "pca.fit(mets_fp)\n",
    "pca_mets_fp = pca.transform(mets_fp)\n",
    "pca_mets_vec_2 = {}\n",
    "for i in range(len(mets_proc)):\n",
    "    k = mets_proc[i]\n",
    "    pca_mets_vec_2[k] = pca_mets_fp[i]"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Get protein vector"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "source": [
    "protein_dir = '../Training_data/pathbank/pathbank_protein_data/'\n",
    "protein_vec = pickle.load(open( protein_dir+\"all_protein_vector.p\", \"rb\" ))"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Get PCA transformed protein feature"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "source": [
    "protein_np = []\n",
    "protein_proc = []\n",
    "for k,v in protein_vec.items():\n",
    "    protein_proc.append(k)\n",
    "    protein_np.append(v)\n",
    "protein_np = np.array(protein_np)\n",
    "protein_np.shape"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "(8291, 1024)"
      ]
     },
     "metadata": {},
     "execution_count": 12
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "source": [
    "# Get PCA transformed data of proteins\n",
    "pca = PCA(n_components=1024)\n",
    "pca.fit(protein_np)\n",
    "pca_protein_np = pca.transform(protein_np)\n",
    "pca_protein_np_1024 = {}\n",
    "for i in range(len(protein_proc)):\n",
    "    k = protein_proc[i]\n",
    "    pca_protein_np_1024[k] = pca_protein_np[i]\n",
    "\n",
    "pca = PCA(n_components=512)\n",
    "pca.fit(protein_np)\n",
    "pca_protein_np = pca.transform(protein_np)\n",
    "pca_protein_np_512 = {}\n",
    "for i in range(len(protein_proc)):\n",
    "    k = protein_proc[i]\n",
    "    pca_protein_np_512[k] = pca_protein_np[i]\n",
    "\n",
    "\n",
    "pca = PCA(n_components=256)\n",
    "pca.fit(protein_np)\n",
    "pca_protein_np = pca.transform(protein_np)\n",
    "pca_protein_np_256 = {}\n",
    "for i in range(len(protein_proc)):\n",
    "    k = protein_proc[i]\n",
    "    pca_protein_np_256[k] = pca_protein_np[i]\n",
    "\n",
    "pca = PCA(n_components=128)\n",
    "pca.fit(protein_np)\n",
    "pca_protein_np = pca.transform(protein_np)\n",
    "pca_protein_np_128 = {}\n",
    "for i in range(len(protein_proc)):\n",
    "    k = protein_proc[i]\n",
    "    pca_protein_np_128[k] = pca_protein_np[i]\n",
    "\n",
    "pca = PCA(n_components=64)\n",
    "pca.fit(protein_np)\n",
    "pca_protein_np = pca.transform(protein_np)\n",
    "pca_protein_np_64 = {}\n",
    "for i in range(len(protein_proc)):\n",
    "    k = protein_proc[i]\n",
    "    pca_protein_np_64[k] = pca_protein_np[i]\n",
    "\n",
    "pca = PCA(n_components=32)\n",
    "pca.fit(protein_np)\n",
    "pca_protein_np = pca.transform(protein_np)\n",
    "pca_protein_np_32 = {}\n",
    "for i in range(len(protein_proc)):\n",
    "    k = protein_proc[i]\n",
    "    pca_protein_np_32[k] = pca_protein_np[i]\n",
    "\n",
    "pca = PCA(n_components=16)\n",
    "pca.fit(protein_np)\n",
    "pca_protein_np = pca.transform(protein_np)\n",
    "pca_protein_np_16 = {}\n",
    "for i in range(len(protein_proc)):\n",
    "    k = protein_proc[i]\n",
    "    pca_protein_np_16[k] = pca_protein_np[i]\n",
    "\n",
    "pca = PCA(n_components=2)\n",
    "pca.fit(protein_np)\n",
    "pca_protein_np = pca.transform(protein_np)\n",
    "pca_protein_np_2 = {}\n",
    "for i in range(len(protein_proc)):\n",
    "    k = protein_proc[i]\n",
    "    pca_protein_np_2[k] = pca_protein_np[i]"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "source": [
    "protein_dir = '../Training_data/pathbank/pathbank_protein_data/'\n",
    "protein_fasta = protein_dir+'pathbank_protein.fasta'"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "source": [
    "with open(protein_fasta) as fasta_file:  \n",
    "    identifiers = {}\n",
    "    for seq_record in SeqIO.parse(fasta_file, 'fasta'):  # (generator)\n",
    "        description = seq_record.description\n",
    "        description = description.split('(')\n",
    "        uniport_id = description[-1][:-1]\n",
    "        identifiers[uniport_id] = (seq_record.seq,len(seq_record.seq))"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "source": [
    "len(identifiers.keys())"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "8291"
      ]
     },
     "metadata": {},
     "execution_count": 16
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Get species specific data\n",
    "### Homo sapiens                20620\n",
    "### Mus musculus                12273\n",
    "### Rattus norvegicus           12207\n",
    "### Escherichia coli             3680\n",
    "### Bos taurus                   3196\n",
    "### Pseudomonas aeruginosa       2736\n",
    "### Arabidopsis thaliana         2006\n",
    "### Saccharomyces cerevisiae     1757\n",
    "### Drosophila melanogaster      1713\n",
    "### Caenorhabditis elegans       1683"
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Generation of species MPI network"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "source": [
    "def get_species_mpi_network(species,main_metabolite_data,main_protein_data):\n",
    "    main_spec_mets_data = main_metabolite_data[main_metabolite_data['Species'] == species] \n",
    "    main_spec_protein_data = main_protein_data[main_protein_data['Species'] == species]\n",
    "    spec_mets_data = metabolite_data[metabolite_data['Species'] == species]\n",
    "    spec_mets_data_m = pd.merge(spec_mets_data,meta_data[['SMPDB ID','PW ID']],left_on='PathBank ID', right_on='SMPDB ID', how='left')\n",
    "    spec_protein_data = protein_data[protein_data['Species'] == species]\n",
    "    spec_protein_data_m = pd.merge(spec_protein_data,meta_data[['SMPDB ID','PW ID']],left_on='PathBank ID', right_on='SMPDB ID', how='left')\n",
    "    'Get metabolite and protein involved pathway ID'\n",
    "    spec_pw_id_met = list(spec_mets_data_m['PW ID'].unique())\n",
    "    spec_pw_id_protn = list(spec_protein_data_m['PW ID'].unique())\n",
    "    spec_pw_id = list(set(spec_pw_id_met+spec_pw_id_protn))\n",
    "    spec_pw_id = [x for x in spec_pw_id if str(x) != 'nan']\n",
    "    g = nx.empty_graph(0, create_using=nx.Graph)\n",
    "    for pwid in spec_pw_id:\n",
    "        pwfile = '../Training_data/pathbank/pathbank_pathway_csv/'+pwid+'*.csv'\n",
    "        try:\n",
    "            spwfile = glob.glob(pwfile)\n",
    "            for spw in spwfile:\n",
    "                df = pd.read_csv(spw,index_col = 0)\n",
    "                g_df = nx.from_pandas_adjacency(df,create_using = nx.Graph())\n",
    "                g = nx.compose(g, g_df)\n",
    "        except:\n",
    "            pass\n",
    "    return g"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "source": [
    "meta_name_id_dict = dict(zip(metabolite_data['Metabolite Name'].tolist(),metabolite_data['Metabolite ID'].tolist()))\n",
    "protein_name_id_dict = dict(zip(protein_data['Protein Name'].tolist(),protein_data['Uniprot ID'].tolist()))"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Generate features for MPI network"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "source": [
    "def get_mpi_feature(g):\n",
    "    all_nodes_df = pd.DataFrame(columns=['node','dbid','class','length','features','pca_1024','pca_512','pca_256','pca_128','pca_64','pca_32','pca_16','pca_2'])\n",
    "    g_nodes = list(g.nodes())\n",
    "    row = [[]]*len(list(all_nodes_df))\n",
    "    for i in range(len(g_nodes)):\n",
    "        node = g_nodes[i]\n",
    "        if node in meta_name_id_dict:\n",
    "            name_id = meta_name_id_dict[node]\n",
    "            row[0:3] = [node, name_id, 'metabolite']\n",
    "            if mets_vec[name_id]:\n",
    "                row[3:] = [mets_vec[name_id][1],mets_vec[name_id][0],pca_mets_vec_1024[name_id],pca_mets_vec_512[name_id],pca_mets_vec_256[name_id],pca_mets_vec_128[name_id],pca_mets_vec_64[name_id],pca_mets_vec_32[name_id],pca_mets_vec_16[name_id],pca_mets_vec_2[name_id]]\n",
    "            else:\n",
    "                row[3:] = [0,np.random.rand((1024)),np.random.rand((1024)),np.random.rand((512)),np.random.rand((256)),np.random.rand((128)),np.random.rand((64)),np.random.rand((32)),np.random.rand((16)),np.random.rand((2))]\n",
    "        elif node in protein_name_id_dict:\n",
    "            name_id = protein_name_id_dict[node]\n",
    "            row[0:3] = [node, name_id, 'protein']\n",
    "            if name_id in protein_vec:\n",
    "                row[3:] = [identifiers[name_id][1],protein_vec[name_id],pca_protein_np_1024[name_id],pca_protein_np_512[name_id],pca_protein_np_256[name_id],pca_protein_np_128[name_id],pca_protein_np_64[name_id],pca_protein_np_32[name_id],pca_protein_np_16[name_id],pca_protein_np_2[name_id]]\n",
    "            else:\n",
    "                row[3:] = [0,np.random.rand((1024)),np.random.rand((1024)),np.random.rand((512)),np.random.rand((256)),np.random.rand((128)),np.random.rand((64)),np.random.rand((32)),np.random.rand((16)),np.random.rand((2))]\n",
    "        else:\n",
    "            row[0:3] = [node, None, 'other']\n",
    "            row[3:] = [0,np.random.rand((1024)),np.random.rand((1024)),np.random.rand((512)),np.random.rand((256)),np.random.rand((128)),np.random.rand((64)),np.random.rand((32)),np.random.rand((16)),np.random.rand((2))]\n",
    "        all_nodes_df.loc[i,:] = row\n",
    "    return all_nodes_df"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "source": [
    "### Homo sapiens                20620\n",
    "### Mus musculus                12273\n",
    "### Rattus norvegicus           12207\n",
    "### Escherichia coli             3680\n",
    "### Bos taurus                   3196\n",
    "### Pseudomonas aeruginosa       2736\n",
    "### Arabidopsis thaliana         2006\n",
    "### Saccharomyces cerevisiae     1757\n",
    "### Drosophila melanogaster      1713\n",
    "### Caenorhabditis elegans       1683\n",
    "species = ['Homo sapiens','Mus musculus','Rattus norvegicus','Escherichia coli','Bos taurus','Pseudomonas aeruginosa',\n",
    "'Arabidopsis thaliana','Saccharomyces cerevisiae','Drosophila melanogaster','Caenorhabditis elegans']\n",
    "for sp in species:\n",
    "    mpi = get_species_mpi_network(sp,main_metabolite_data,main_protein_data)\n",
    "    nodes_df = get_mpi_feature(mpi)\n",
    "    mpi_file_name = '../features/mpi_network/'+'pca_mpi_'+str(sp).replace(' ','_')+'.pkl'\n",
    "    df_file_name = '../features/mpi_features/'+'pca_feature_df_'+str(sp).replace(' ','_')+'.pkl'\n",
    "    mpi_data_file = open(mpi_file_name, 'wb') \n",
    "    pickle.dump(mpi,mpi_data_file)\n",
    "    df_file = open(df_file_name,'wb')\n",
    "    pickle.dump(nodes_df,df_file)"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "/Users/chengwang/opt/anaconda3/lib/python3.8/site-packages/pandas/core/internals/blocks.py:993: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  arr_value = np.array(value)\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "source": [
    "def get_species_mpi_network_directed(species,main_metabolite_data,main_protein_data):\n",
    "    main_spec_mets_data = main_metabolite_data[main_metabolite_data['Species'] == species] \n",
    "    main_spec_protein_data = main_protein_data[main_protein_data['Species'] == species]\n",
    "    spec_mets_data = metabolite_data[metabolite_data['Species'] == species]\n",
    "    spec_mets_data_m = pd.merge(spec_mets_data,meta_data[['SMPDB ID','PW ID']],left_on='PathBank ID', right_on='SMPDB ID', how='left')\n",
    "    spec_protein_data = protein_data[protein_data['Species'] == species]\n",
    "    spec_protein_data_m = pd.merge(spec_protein_data,meta_data[['SMPDB ID','PW ID']],left_on='PathBank ID', right_on='SMPDB ID', how='left')\n",
    "    'Get metabolite and protein involved pathway ID'\n",
    "    spec_pw_id_met = list(spec_mets_data_m['PW ID'].unique())\n",
    "    spec_pw_id_protn = list(spec_protein_data_m['PW ID'].unique())\n",
    "    spec_pw_id = list(set(spec_pw_id_met+spec_pw_id_protn))\n",
    "    spec_pw_id = [x for x in spec_pw_id if str(x) != 'nan']\n",
    "    g = nx.empty_graph(0, create_using=nx.Graph)\n",
    "    for pwid in spec_pw_id:\n",
    "        pwfile = '../Training_data/pathbank/pathbank_pathway_csv/'+pwid+'*.csv'\n",
    "        try:\n",
    "            spwfile = glob.glob(pwfile)\n",
    "            for spw in spwfile:\n",
    "                df = pd.read_csv(spw,index_col = 0)\n",
    "                g_df = nx.from_pandas_adjacency(df,create_using = nx.DiGraph())\n",
    "                g = nx.compose(g, g_df)\n",
    "        except:\n",
    "            pass\n",
    "    return g"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "source": [
    "### Homo sapiens                20620\n",
    "### Mus musculus                12273\n",
    "### Rattus norvegicus           12207\n",
    "### Escherichia coli             3680\n",
    "### Bos taurus                   3196\n",
    "### Pseudomonas aeruginosa       2736\n",
    "### Arabidopsis thaliana         2006\n",
    "### Saccharomyces cerevisiae     1757\n",
    "### Drosophila melanogaster      1713\n",
    "### Caenorhabditis elegans       1683\n",
    "species = ['Homo sapiens','Mus musculus','Rattus norvegicus','Escherichia coli','Bos taurus','Pseudomonas aeruginosa',\n",
    "'Arabidopsis thaliana','Saccharomyces cerevisiae','Drosophila melanogaster','Caenorhabditis elegans']\n",
    "for sp in species[:1]:\n",
    "    mpi = get_species_mpi_network_directed(sp,main_metabolite_data,main_protein_data)"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "source": [],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "source": [
    "print(nx.info(mpi))"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "Name: \n",
      "Type: Graph\n",
      "Number of nodes: 2306\n",
      "Number of edges: 5785\n",
      "Average degree:   5.0173\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "source": [
    "species = ['Homo sapiens','Escherichia coli','Bos taurus','Pseudomonas aeruginosa','Arabidopsis thaliana']\n",
    "seed_number = [12345, 22345, 32345, 42345, 52345]\n",
    "index = 0\n",
    "result = {}\n",
    "for spcs in species[:1]:\n",
    "    spcs_temp = {}\n",
    "    for seed_num in seed_number[:1]:\n",
    "        print(spcs,seed_num)\n",
    "        # row = [[]]*len(list(result_charc))\n",
    "        mpi_file_name = '../features/mpi_network/'+'mpi_'+str(spcs).replace(' ','_')+'.pkl'\n",
    "        df_file_name = '../features/mpi_features/'+'feature_df_'+str(spcs).replace(' ','_')+'.pkl'\n",
    "        g = pickle.load(open(mpi_file_name, \"rb\" ))\n",
    "        node_feats = pickle.load(open(df_file_name, \"rb\" ))"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "Homo sapiens 12345\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "source": [
    "mets_id_select = metabolite_data[['Metabolite Name','Metabolite ID','HMDB ID','KEGG ID']]\n",
    "mets_id_select = mets_id_select.drop_duplicates()\n",
    "protn_id_select = protein_data[['Protein Name','Uniprot ID','Gene Name']]\n",
    "protn_id_select = protn_id_select.drop_duplicates()"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "source": [
    "node_feats[node_feats['node'] == 'Glucose-6-phosphate isomerase']"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node</th>\n",
       "      <th>dbid</th>\n",
       "      <th>class</th>\n",
       "      <th>length</th>\n",
       "      <th>features</th>\n",
       "      <th>pca_1024</th>\n",
       "      <th>pca_512</th>\n",
       "      <th>pca_256</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>728</th>\n",
       "      <td>Glucose-6-phosphate isomerase</td>\n",
       "      <td>A0A072ZX40</td>\n",
       "      <td>protein</td>\n",
       "      <td>554</td>\n",
       "      <td>[0.053579077, -0.013836206, -0.013869764, -0.0...</td>\n",
       "      <td>[-0.3121425, -0.2838221, 0.60420144, -0.609301...</td>\n",
       "      <td>[-0.31214222, -0.2838224, 0.60420084, -0.60930...</td>\n",
       "      <td>[-0.31214264, -0.2838218, 0.60420203, -0.60930...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              node        dbid    class length  \\\n",
       "728  Glucose-6-phosphate isomerase  A0A072ZX40  protein    554   \n",
       "\n",
       "                                              features  \\\n",
       "728  [0.053579077, -0.013836206, -0.013869764, -0.0...   \n",
       "\n",
       "                                              pca_1024  \\\n",
       "728  [-0.3121425, -0.2838221, 0.60420144, -0.609301...   \n",
       "\n",
       "                                               pca_512  \\\n",
       "728  [-0.31214222, -0.2838224, 0.60420084, -0.60930...   \n",
       "\n",
       "                                               pca_256  \n",
       "728  [-0.31214264, -0.2838218, 0.60420203, -0.60930...  "
      ]
     },
     "metadata": {},
     "execution_count": 142
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "source": [
    "node_feats = node_feats.merge(mets_id_select,left_on='dbid',right_on='Metabolite ID',how='left')\n",
    "print(len(node_feats))\n",
    "node_feats = node_feats.merge(protn_id_select,left_on='dbid',right_on='Uniprot ID',how='left')\n",
    "print(len(node_feats))"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "2306\n",
      "4968\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "source": [
    "node_feats.drop_duplicates(subset='node', keep=\"last\",inplace=True)\n",
    "len(node_feats)"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "2306"
      ]
     },
     "metadata": {},
     "execution_count": 144
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "source": [
    "protein_list = list(set(node_feats['Gene Name'].tolist()))[1:]\n",
    "proteins = '%0d'.join(protein_list)\n",
    "url = 'https://string-db.org/api/tsv/network?identifiers=' + proteins + '&species=9606'\n",
    "r = requests.get(url)\n",
    "lines = r.text.split('\\n') # pull the text from the response object and split based on new lines\n",
    "data = [l.split('\\t') for l in lines] # split each line into its components based on tabs\n",
    "# convert to dataframe using the first row as the column names; drop empty, final row\n",
    "df = pd.DataFrame(data[1:-1], columns = data[0]) \n",
    "# dataframe with the preferred names of the two proteins and the score of the interaction\n",
    "interactions = df[['preferredName_A', 'preferredName_B', 'score']]"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "source": [
    "g_kg = nx.empty_graph(0, create_using=nx.Graph)\n",
    "\n",
    "adj_df = pd.read_csv('../features/kegg/kegg_reactions.csv',index_col = 0)\n",
    "g_df = nx.from_pandas_adjacency(adj_df,create_using = nx.Graph())\n",
    "g_kg = nx.compose(g_kg, g_df)"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "source": [
    "kegg_df_edge = pd.DataFrame(columns=['kegg_id_1','kegg_id_2'])\n",
    "index = 0\n",
    "for edge in list(g_kg.edges()):\n",
    "    edge_1 = edge[0].replace('cpd:','')\n",
    "    edge_2 = edge[1].replace('cpd:','')\n",
    "    kegg_df_edge.loc[index,:] = [edge_1,edge_2]\n",
    "    index += 1"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "source": [
    "node_feats.head()"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node</th>\n",
       "      <th>dbid</th>\n",
       "      <th>class</th>\n",
       "      <th>length</th>\n",
       "      <th>features</th>\n",
       "      <th>pca_1024</th>\n",
       "      <th>pca_512</th>\n",
       "      <th>pca_256</th>\n",
       "      <th>Metabolite Name</th>\n",
       "      <th>Metabolite ID</th>\n",
       "      <th>HMDB ID</th>\n",
       "      <th>KEGG ID</th>\n",
       "      <th>Protein Name</th>\n",
       "      <th>Uniprot ID</th>\n",
       "      <th>Gene Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Enoyl-CoA hydratase, mitochondrial</td>\n",
       "      <td>P14604</td>\n",
       "      <td>protein</td>\n",
       "      <td>290</td>\n",
       "      <td>[-0.0115655055, -0.02298307, -0.023107061, 0.0...</td>\n",
       "      <td>[0.69804204, 1.5457662, -0.19488898, -0.072530...</td>\n",
       "      <td>[0.69804233, 1.5457658, -0.1948879, -0.0725309...</td>\n",
       "      <td>[0.6980426, 1.5457666, -0.19489, -0.0725313, -...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Enoyl-CoA hydratase, mitochondrial</td>\n",
       "      <td>P14604</td>\n",
       "      <td>Echs1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3-Hydroxybutyryl-CoA</td>\n",
       "      <td>PW_C000904</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>54</td>\n",
       "      <td>[0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, ...</td>\n",
       "      <td>[0.4531775710572908, 1.3835354738819938, 5.772...</td>\n",
       "      <td>[0.45317757105729123, 1.38353547388199, 5.7727...</td>\n",
       "      <td>[0.45317757105729006, 1.383535473881945, 5.772...</td>\n",
       "      <td>3-Hydroxybutyryl-CoA</td>\n",
       "      <td>PW_C000904</td>\n",
       "      <td>HMDB0001166</td>\n",
       "      <td>C03561</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Crotonoyl-CoA</td>\n",
       "      <td>PW_C001345</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>53</td>\n",
       "      <td>[0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, ...</td>\n",
       "      <td>[0.04700508297648426, 1.1281863993703316, 5.78...</td>\n",
       "      <td>[0.04700508297648365, 1.1281863993703287, 5.78...</td>\n",
       "      <td>[0.047005082976483314, 1.1281863993702816, 5.7...</td>\n",
       "      <td>Crotonoyl-CoA</td>\n",
       "      <td>PW_C001345</td>\n",
       "      <td>HMDB0002009</td>\n",
       "      <td>C00877</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Water</td>\n",
       "      <td>PW_C001420</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>1</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[-0.16956676756781652, 2.8263338162518443, 3.2...</td>\n",
       "      <td>[-0.1695667675678163, 2.8263338162518394, 3.22...</td>\n",
       "      <td>[-0.1695667675678263, 2.8263338162518137, 3.22...</td>\n",
       "      <td>Water</td>\n",
       "      <td>PW_C001420</td>\n",
       "      <td>HMDB0002111</td>\n",
       "      <td>C00001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Acetyl-CoA acetyltransferase, mitochondrial</td>\n",
       "      <td>P24752</td>\n",
       "      <td>protein</td>\n",
       "      <td>427</td>\n",
       "      <td>[-0.10862053, -0.12662861, -0.16699015, 0.1391...</td>\n",
       "      <td>[0.66844195, 1.3619962, -0.062943734, -0.75548...</td>\n",
       "      <td>[0.66844237, 1.3619953, -0.062942795, -0.75548...</td>\n",
       "      <td>[0.6684426, 1.3619963, -0.062944755, -0.755485...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Acetyl-CoA acetyltransferase, mitochondrial</td>\n",
       "      <td>P24752</td>\n",
       "      <td>ACAT1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          node        dbid       class length  \\\n",
       "0           Enoyl-CoA hydratase, mitochondrial      P14604     protein    290   \n",
       "1                         3-Hydroxybutyryl-CoA  PW_C000904  metabolite     54   \n",
       "2                                Crotonoyl-CoA  PW_C001345  metabolite     53   \n",
       "3                                        Water  PW_C001420  metabolite      1   \n",
       "4  Acetyl-CoA acetyltransferase, mitochondrial      P24752     protein    427   \n",
       "\n",
       "                                            features  \\\n",
       "0  [-0.0115655055, -0.02298307, -0.023107061, 0.0...   \n",
       "1  [0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, ...   \n",
       "2  [0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, ...   \n",
       "3  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "4  [-0.10862053, -0.12662861, -0.16699015, 0.1391...   \n",
       "\n",
       "                                            pca_1024  \\\n",
       "0  [0.69804204, 1.5457662, -0.19488898, -0.072530...   \n",
       "1  [0.4531775710572908, 1.3835354738819938, 5.772...   \n",
       "2  [0.04700508297648426, 1.1281863993703316, 5.78...   \n",
       "3  [-0.16956676756781652, 2.8263338162518443, 3.2...   \n",
       "4  [0.66844195, 1.3619962, -0.062943734, -0.75548...   \n",
       "\n",
       "                                             pca_512  \\\n",
       "0  [0.69804233, 1.5457658, -0.1948879, -0.0725309...   \n",
       "1  [0.45317757105729123, 1.38353547388199, 5.7727...   \n",
       "2  [0.04700508297648365, 1.1281863993703287, 5.78...   \n",
       "3  [-0.1695667675678163, 2.8263338162518394, 3.22...   \n",
       "4  [0.66844237, 1.3619953, -0.062942795, -0.75548...   \n",
       "\n",
       "                                             pca_256       Metabolite Name  \\\n",
       "0  [0.6980426, 1.5457666, -0.19489, -0.0725313, -...                   NaN   \n",
       "1  [0.45317757105729006, 1.383535473881945, 5.772...  3-Hydroxybutyryl-CoA   \n",
       "2  [0.047005082976483314, 1.1281863993702816, 5.7...         Crotonoyl-CoA   \n",
       "3  [-0.1695667675678263, 2.8263338162518137, 3.22...                 Water   \n",
       "4  [0.6684426, 1.3619963, -0.062944755, -0.755485...                   NaN   \n",
       "\n",
       "  Metabolite ID      HMDB ID KEGG ID  \\\n",
       "0           NaN          NaN     NaN   \n",
       "1    PW_C000904  HMDB0001166  C03561   \n",
       "2    PW_C001345  HMDB0002009  C00877   \n",
       "3    PW_C001420  HMDB0002111  C00001   \n",
       "4           NaN          NaN     NaN   \n",
       "\n",
       "                                  Protein Name Uniprot ID Gene Name  \n",
       "0           Enoyl-CoA hydratase, mitochondrial     P14604     Echs1  \n",
       "1                                          NaN        NaN       NaN  \n",
       "2                                          NaN        NaN       NaN  \n",
       "3                                          NaN        NaN       NaN  \n",
       "4  Acetyl-CoA acetyltransferase, mitochondrial     P24752     ACAT1  "
      ]
     },
     "metadata": {},
     "execution_count": 149
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "source": [
    "'add edges'\n",
    "\"g.edges()\"\n",
    "'gene name : protein'\n",
    "'kegg id': 'metabolite name'"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "'g.edges()'"
      ]
     },
     "metadata": {},
     "execution_count": 89
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "source": [
    "gene_name_protein_dict = dict(zip(node_feats['Gene Name'].tolist(),node_feats['node'].tolist()))\n",
    "kegg_id_metabolite_dict = dict(zip(node_feats['KEGG ID'].tolist(),node_feats['node'].tolist()))"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "source": [
    "interactions.drop_duplicates(inplace=True)"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "<ipython-input-151-edea3f19db32>:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  interactions.drop_duplicates(inplace=True)\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "source": [
    "len(interactions)"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "12751"
      ]
     },
     "metadata": {},
     "execution_count": 152
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "source": [
    "for index, row in interactions.iterrows():\n",
    "    node_1 = row['preferredName_A']\n",
    "    node_2 = row['preferredName_B']\n",
    "    if node_1 in gene_name_protein_dict and node_2 in gene_name_protein_dict:\n",
    "        n1 = gene_name_protein_dict[node_1]\n",
    "        n2 = gene_name_protein_dict[node_2]\n",
    "        # print((node_1,node_2))\n",
    "        ppi_edge = (node_1,node_2)\n",
    "        g.add_edge(n1, n2)"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "source": [
    "kegg_df_edge.head()"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>kegg_id_1</th>\n",
       "      <th>kegg_id_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C00084</td>\n",
       "      <td>C00024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C00084</td>\n",
       "      <td>C05125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C00084</td>\n",
       "      <td>C00469</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C00084</td>\n",
       "      <td>C00186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C00084</td>\n",
       "      <td>C00033</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  kegg_id_1 kegg_id_2\n",
       "0    C00084    C00024\n",
       "1    C00084    C05125\n",
       "2    C00084    C00469\n",
       "3    C00084    C00186\n",
       "4    C00084    C00033"
      ]
     },
     "metadata": {},
     "execution_count": 154
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "source": [
    "counts = 0\n",
    "for index, row in kegg_df_edge.iterrows():\n",
    "    node_1 = row['kegg_id_1']\n",
    "    node_2 = row['kegg_id_2']\n",
    "    if node_1 in kegg_id_metabolite_dict and node_2 in kegg_id_metabolite_dict:\n",
    "        n1 = kegg_id_metabolite_dict[node_1]\n",
    "        n2 = kegg_id_metabolite_dict[node_2]\n",
    "        counts += 1\n",
    "        # print((node_1,node_2))\n",
    "        g.add_edge(n1, n2)"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "source": [
    "print(nx.info(g))"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "Name: \n",
      "Type: Graph\n",
      "Number of nodes: 2306\n",
      "Number of edges: 8500\n",
      "Average degree:   7.3721\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "source": [
    "len(g.nodes())"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "2306"
      ]
     },
     "metadata": {},
     "execution_count": 157
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "source": [
    "node_feats.head()"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node</th>\n",
       "      <th>dbid</th>\n",
       "      <th>class</th>\n",
       "      <th>length</th>\n",
       "      <th>features</th>\n",
       "      <th>pca_1024</th>\n",
       "      <th>pca_512</th>\n",
       "      <th>pca_256</th>\n",
       "      <th>Metabolite Name</th>\n",
       "      <th>Metabolite ID</th>\n",
       "      <th>HMDB ID</th>\n",
       "      <th>KEGG ID</th>\n",
       "      <th>Protein Name</th>\n",
       "      <th>Uniprot ID</th>\n",
       "      <th>Gene Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Enoyl-CoA hydratase, mitochondrial</td>\n",
       "      <td>P14604</td>\n",
       "      <td>protein</td>\n",
       "      <td>290</td>\n",
       "      <td>[-0.0115655055, -0.02298307, -0.023107061, 0.0...</td>\n",
       "      <td>[0.69804204, 1.5457662, -0.19488898, -0.072530...</td>\n",
       "      <td>[0.69804233, 1.5457658, -0.1948879, -0.0725309...</td>\n",
       "      <td>[0.6980426, 1.5457666, -0.19489, -0.0725313, -...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Enoyl-CoA hydratase, mitochondrial</td>\n",
       "      <td>P14604</td>\n",
       "      <td>Echs1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3-Hydroxybutyryl-CoA</td>\n",
       "      <td>PW_C000904</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>54</td>\n",
       "      <td>[0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, ...</td>\n",
       "      <td>[0.4531775710572908, 1.3835354738819938, 5.772...</td>\n",
       "      <td>[0.45317757105729123, 1.38353547388199, 5.7727...</td>\n",
       "      <td>[0.45317757105729006, 1.383535473881945, 5.772...</td>\n",
       "      <td>3-Hydroxybutyryl-CoA</td>\n",
       "      <td>PW_C000904</td>\n",
       "      <td>HMDB0001166</td>\n",
       "      <td>C03561</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Crotonoyl-CoA</td>\n",
       "      <td>PW_C001345</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>53</td>\n",
       "      <td>[0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, ...</td>\n",
       "      <td>[0.04700508297648426, 1.1281863993703316, 5.78...</td>\n",
       "      <td>[0.04700508297648365, 1.1281863993703287, 5.78...</td>\n",
       "      <td>[0.047005082976483314, 1.1281863993702816, 5.7...</td>\n",
       "      <td>Crotonoyl-CoA</td>\n",
       "      <td>PW_C001345</td>\n",
       "      <td>HMDB0002009</td>\n",
       "      <td>C00877</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Water</td>\n",
       "      <td>PW_C001420</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>1</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[-0.16956676756781652, 2.8263338162518443, 3.2...</td>\n",
       "      <td>[-0.1695667675678163, 2.8263338162518394, 3.22...</td>\n",
       "      <td>[-0.1695667675678263, 2.8263338162518137, 3.22...</td>\n",
       "      <td>Water</td>\n",
       "      <td>PW_C001420</td>\n",
       "      <td>HMDB0002111</td>\n",
       "      <td>C00001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Acetyl-CoA acetyltransferase, mitochondrial</td>\n",
       "      <td>P24752</td>\n",
       "      <td>protein</td>\n",
       "      <td>427</td>\n",
       "      <td>[-0.10862053, -0.12662861, -0.16699015, 0.1391...</td>\n",
       "      <td>[0.66844195, 1.3619962, -0.062943734, -0.75548...</td>\n",
       "      <td>[0.66844237, 1.3619953, -0.062942795, -0.75548...</td>\n",
       "      <td>[0.6684426, 1.3619963, -0.062944755, -0.755485...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Acetyl-CoA acetyltransferase, mitochondrial</td>\n",
       "      <td>P24752</td>\n",
       "      <td>ACAT1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          node        dbid       class length  \\\n",
       "0           Enoyl-CoA hydratase, mitochondrial      P14604     protein    290   \n",
       "1                         3-Hydroxybutyryl-CoA  PW_C000904  metabolite     54   \n",
       "2                                Crotonoyl-CoA  PW_C001345  metabolite     53   \n",
       "3                                        Water  PW_C001420  metabolite      1   \n",
       "4  Acetyl-CoA acetyltransferase, mitochondrial      P24752     protein    427   \n",
       "\n",
       "                                            features  \\\n",
       "0  [-0.0115655055, -0.02298307, -0.023107061, 0.0...   \n",
       "1  [0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, ...   \n",
       "2  [0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, ...   \n",
       "3  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "4  [-0.10862053, -0.12662861, -0.16699015, 0.1391...   \n",
       "\n",
       "                                            pca_1024  \\\n",
       "0  [0.69804204, 1.5457662, -0.19488898, -0.072530...   \n",
       "1  [0.4531775710572908, 1.3835354738819938, 5.772...   \n",
       "2  [0.04700508297648426, 1.1281863993703316, 5.78...   \n",
       "3  [-0.16956676756781652, 2.8263338162518443, 3.2...   \n",
       "4  [0.66844195, 1.3619962, -0.062943734, -0.75548...   \n",
       "\n",
       "                                             pca_512  \\\n",
       "0  [0.69804233, 1.5457658, -0.1948879, -0.0725309...   \n",
       "1  [0.45317757105729123, 1.38353547388199, 5.7727...   \n",
       "2  [0.04700508297648365, 1.1281863993703287, 5.78...   \n",
       "3  [-0.1695667675678163, 2.8263338162518394, 3.22...   \n",
       "4  [0.66844237, 1.3619953, -0.062942795, -0.75548...   \n",
       "\n",
       "                                             pca_256       Metabolite Name  \\\n",
       "0  [0.6980426, 1.5457666, -0.19489, -0.0725313, -...                   NaN   \n",
       "1  [0.45317757105729006, 1.383535473881945, 5.772...  3-Hydroxybutyryl-CoA   \n",
       "2  [0.047005082976483314, 1.1281863993702816, 5.7...         Crotonoyl-CoA   \n",
       "3  [-0.1695667675678263, 2.8263338162518137, 3.22...                 Water   \n",
       "4  [0.6684426, 1.3619963, -0.062944755, -0.755485...                   NaN   \n",
       "\n",
       "  Metabolite ID      HMDB ID KEGG ID  \\\n",
       "0           NaN          NaN     NaN   \n",
       "1    PW_C000904  HMDB0001166  C03561   \n",
       "2    PW_C001345  HMDB0002009  C00877   \n",
       "3    PW_C001420  HMDB0002111  C00001   \n",
       "4           NaN          NaN     NaN   \n",
       "\n",
       "                                  Protein Name Uniprot ID Gene Name  \n",
       "0           Enoyl-CoA hydratase, mitochondrial     P14604     Echs1  \n",
       "1                                          NaN        NaN       NaN  \n",
       "2                                          NaN        NaN       NaN  \n",
       "3                                          NaN        NaN       NaN  \n",
       "4  Acetyl-CoA acetyltransferase, mitochondrial     P24752     ACAT1  "
      ]
     },
     "metadata": {},
     "execution_count": 115
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "source": [
    "len(node_feats)"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "2306"
      ]
     },
     "metadata": {},
     "execution_count": 158
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "source": [
    "mpi_file_name = '../features/mpi_network/'+'mpi_hs_all.pkl'\n",
    "df_file_name = '../features/mpi_features/'+'feature_df_mpi_hs_all.pkl'\n",
    "mpi_data_file = open(mpi_file_name, 'wb') \n",
    "pickle.dump(g,mpi_data_file)\n",
    "df_file = open(df_file_name,'wb')\n",
    "pickle.dump(node_feats,df_file)"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "## MPI network description"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "source": [
    "mpi_data_file = open('../features/mpi_network/mpi_Homo_sapiens.pkl', 'rb') \n",
    "g = pickle.load(mpi_data_file)"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "source": [
    "nodes = list(g.nodes())"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "source": [
    "c_mets = 0\n",
    "c_prtn = 0\n",
    "c_prtn_nodes = []\n",
    "c_mets_nodes = []\n",
    "for i in nodes:\n",
    "    if i in meta_name_id_dict:\n",
    "        c_mets += 1\n",
    "        c_mets_nodes.append(i)\n",
    "    if i in protein_name_id_dict:\n",
    "        c_prtn += 1\n",
    "        c_prtn_nodes.append(i)\n",
    "print('number of metabolites ',c_mets)\n",
    "print('number of protein ',c_prtn)\n",
    "print('total number ',len(nodes),c_mets+c_prtn,len(nodes)-c_prtn)"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "number of metabolites  1261\n",
      "number of protein  855\n",
      "total number  2306 2116 1451\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "source": [
    "protein_name_id_dict"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "{'Alanine--tRNA ligase, cytoplasmic': 'P50475',\n",
       " 'Alanine aminotransferase 1': 'Q4V7F7',\n",
       " 'Pyruvate carboxylase, mitochondrial': 'Q64555',\n",
       " 'Serine--pyruvate aminotransferase': 'Q94055',\n",
       " 'Alanine--tRNA ligase, mitochondrial': 'D3ZX08',\n",
       " 'Mitochondrial pyruvate carrier 1': 'Q4V8N5',\n",
       " '4-aminobutyrate aminotransferase, mitochondrial': 'Q66HM1',\n",
       " 'Glutamate decarboxylase 1': 'P18088',\n",
       " 'Asparagine synthetase [glutamine-hydrolyzing]': 'Q66HR8',\n",
       " 'Asparagine--tRNA ligase, cytoplasmic': 'Q8BP47',\n",
       " 'Argininosuccinate synthase': 'V6AB05',\n",
       " 'Aspartate--tRNA ligase, cytoplasmic': 'P15178',\n",
       " 'Adenylosuccinate lyase': 'Q9I0K9',\n",
       " 'D-aspartate oxidase': 'Q922Z0',\n",
       " 'Aspartoacylase': 'Q6AZ03',\n",
       " 'Adenylosuccinate synthetase isozyme 1': 'M0R629',\n",
       " 'CAD protein': 'B7ZN27',\n",
       " 'Isoaspartyl peptidase/L-asparaginase': 'Q8CG44',\n",
       " 'Argininosuccinate lyase': 'Q02EA0',\n",
       " 'L-amino-acid oxidase': 'Q9CXK7',\n",
       " 'Unknown': 'Unknown',\n",
       " 'Glutamine synthetase': 'Q9HU65',\n",
       " 'Glutaminase liver isoform, mitochondrial': 'Q64606',\n",
       " 'Glutamine--fructose-6-phosphate aminotransferase [isomerizing] 1': 'Q5XI08',\n",
       " 'Glucosamine 6-phosphate N-acetyltransferase': 'Q9JK38',\n",
       " 'N-acetyl-D-glucosamine kinase': 'A0A0G5QQU4',\n",
       " 'Carbamoyl-phosphate synthase [ammonia], mitochondrial': 'P07756',\n",
       " 'Amidophosphoribosyltransferase': 'Q9HZB1',\n",
       " 'GMP synthase [glutamine-hydrolyzing]': 'Q02RS2',\n",
       " 'Glutamine-dependent NAD(+) synthetase': 'Q812E8',\n",
       " 'Delta-1-pyrroline-5-carboxylate dehydrogenase, mitochondrial': 'P0C2X9',\n",
       " 'Glutamate dehydrogenase 1, mitochondrial': 'Q6LC16',\n",
       " 'Aspartate aminotransferase, mitochondrial': 'Q9QV50',\n",
       " 'Probable glutamate--tRNA ligase, mitochondrial': 'Q9CXJ1',\n",
       " 'Glutamate--cysteine ligase catalytic subunit': 'P19468',\n",
       " 'Glutamate--cysteine ligase regulatory subunit': 'P48508',\n",
       " 'Glutathione synthetase': 'Q9I697',\n",
       " 'Glutathione reductase, mitochondrial': 'P47791',\n",
       " 'Succinate-semialdehyde dehydrogenase, mitochondrial': 'P51650',\n",
       " 'Glutamine--tRNA ligase': 'Q02KT5',\n",
       " 'Glutathione peroxidase 1': 'Q91WZ5',\n",
       " 'Glutathione S-transferase omega-2': 'Q6AXV9',\n",
       " 'Gamma-glutamyltransferase 6': 'Q6P531',\n",
       " 'Aminopeptidase N': 'Q9JHP4',\n",
       " 'Caspase-7': 'O08669',\n",
       " '5-oxoprolinase': 'Q5U3Z8',\n",
       " 'Gamma-glutamylcyclotransferase': 'Q66JQ5',\n",
       " 'Pyruvate dehydrogenase E1 component subunit alpha, somatic form, mitochondrial': 'P26284',\n",
       " 'Pyruvate dehydrogenase E1 component subunit beta, mitochondrial': 'O44451',\n",
       " 'Dihydrolipoyllysine-residue acetyltransferase component of pyruvate dehydrogenase complex, mitochondrial': 'Q19749',\n",
       " 'Dihydrolipoyl dehydrogenase, mitochondrial': 'D3YT81',\n",
       " 'Citrate synthase, mitochondrial': 'Q8VHF5',\n",
       " 'Isocitrate dehydrogenase [NAD] subunit alpha, mitochondrial': 'Q99NA5',\n",
       " 'Isocitrate dehydrogenase [NAD] subunit beta, mitochondrial': 'Q68FX0',\n",
       " 'Isocitrate dehydrogenase [NAD] subunit gamma, mitochondrial': 'Q148J3',\n",
       " '2-oxoglutarate dehydrogenase, mitochondrial': 'O61199',\n",
       " 'Dihydrolipoyllysine-residue succinyltransferase component of 2-oxoglutarate dehydrogenase complex, mitochondrial': 'P36957',\n",
       " 'Succinyl-CoA ligase [ADP/GDP-forming] subunit alpha, mitochondrial': 'P53597',\n",
       " 'Succinyl-CoA ligase [GDP-forming] subunit beta, mitochondrial': 'Q96I99',\n",
       " 'Fumarate hydratase, mitochondrial': 'P14408',\n",
       " 'Malate dehydrogenase, mitochondrial': 'Q6GSM4',\n",
       " 'Succinate dehydrogenase [ubiquinone] flavoprotein subunit, mitochondrial': 'Q09508',\n",
       " 'Succinate dehydrogenase [ubiquinone] iron-sulfur subunit, mitochondrial': 'Q0QEZ3',\n",
       " 'Succinate dehydrogenase cytochrome b560 subunit, mitochondrial': 'Q9T2T6',\n",
       " 'Succinate dehydrogenase [ubiquinone] cytochrome b small subunit, mitochondrial': 'Q6PCT8',\n",
       " 'Aconitate hydratase, mitochondrial': 'Q6P6V3',\n",
       " 'Fatty acid desaturase 2': 'H2BF31',\n",
       " 'Elongation of very long chain fatty acids protein 5': 'G9BD46',\n",
       " 'Fatty acid desaturase 1': 'Q9EPV4',\n",
       " 'Elongation of very long chain fatty acids protein 4': 'Q8JZV3',\n",
       " 'Cytosolic phospholipase A2 beta': 'P0C871',\n",
       " 'Selenocysteine lyase': 'Q68FT9',\n",
       " 'Selenide, water dikinase 2': 'Q80ZY9',\n",
       " 'Cystathionine gamma-lyase': 'P18757',\n",
       " 'Cystathionine beta-synthase': 'P32232',\n",
       " 'Adenosylhomocysteinase': 'P10760',\n",
       " 'Methyltransferase-like protein 6': 'Q6AXU8',\n",
       " 'S-adenosylmethionine synthase isoform type-2': 'P18298',\n",
       " 'Methionine adenosyltransferase 2 subunit beta': 'Q5U2R0',\n",
       " 'Methionine--tRNA ligase, cytoplasmic': 'Q68FL6',\n",
       " \"Bifunctional 3'-phosphoadenosine 5'-phosphosulfate synthase 1\": 'Q3U647',\n",
       " 'Gamma-glutamyltranspeptidase 1': 'P19440',\n",
       " 'Zinc transporter ZIP8': 'Q9C0K1',\n",
       " 'Chitotriosidase-1': 'Q6QJD2',\n",
       " 'Beta-hexosaminidase subunit alpha': 'Q641X3',\n",
       " 'Putative N-acetylglucosamine-6-phosphate deacetylase': 'Q9Y303',\n",
       " 'Glucosamine-6-phosphate isomerase 1': 'Q9R188',\n",
       " 'Phosphoacetylglucosamine mutase': 'B2RYN0',\n",
       " 'UDP-N-acetylhexosamine pyrophosphorylase': 'Q8VD59',\n",
       " 'N-acylglucosamine 2-epimerase': 'P51607',\n",
       " 'Bifunctional UDP-N-acetylglucosamine 2-epimerase/N-acetylmannosamine kinase': 'O35826',\n",
       " 'Hexokinase-1': 'P05708',\n",
       " 'Sialic acid synthase': 'Q9JJH0',\n",
       " 'N-acetylneuraminate lyase': 'Q66H59',\n",
       " 'N-acylneuraminate-9-phosphatase': 'Q5M969',\n",
       " 'N-acylneuraminate cytidylyltransferase': 'P69060',\n",
       " 'Sialin': 'Q3TE25',\n",
       " 'L-serine dehydratase/L-threonine deaminase': 'Q5M8C4',\n",
       " 'Histidine ammonia-lyase': 'Q5EBB8',\n",
       " 'Aminomethyltransferase, mitochondrial': 'Q8CFA2',\n",
       " 'Glycine dehydrogenase [decarboxylating], mitochondrial': 'P23378',\n",
       " 'Glycine cleavage system H protein, mitochondrial': 'Q9VP59',\n",
       " 'Aquaporin-8': 'O94778',\n",
       " 'Ornithine aminotransferase, mitochondrial': 'Q6LDF7',\n",
       " 'Proline dehydrogenase 1, mitochondrial': 'Q9WU79',\n",
       " 'Pyrroline-5-carboxylate reductase 2': 'Q6AY23',\n",
       " 'Prolyl 4-hydroxylase subunit alpha-3': 'Q6W3E9',\n",
       " 'Probable arginine--tRNA ligase, mitochondrial': 'Q3U186',\n",
       " 'Glycine amidinotransferase, mitochondrial': 'Q6ITZ7',\n",
       " 'Guanidinoacetate N-methyltransferase': 'P10868',\n",
       " 'Creatine kinase B-type': 'Q6P139',\n",
       " 'Arginase-1': 'Q5BK93',\n",
       " 'Ornithine carbamoyltransferase, mitochondrial': 'Q63407',\n",
       " 'Mitochondrial ornithine transporter 1': 'A0A0G2K309',\n",
       " 'D-amino-acid oxidase': 'O35078',\n",
       " 'Aspartate aminotransferase, cytoplasmic': 'Q6P721',\n",
       " 'Nitric oxide synthase, brain': 'P70594',\n",
       " 'Bifunctional glutamate/proline--tRNA ligase': 'Q8CGC7',\n",
       " 'Beta-Ala-His dipeptidase': 'Q66HG3',\n",
       " 'Beta-ureidopropionase': 'Q03248',\n",
       " 'Dihydropyrimidinase': 'Q642F0',\n",
       " 'Dihydropyrimidine dehydrogenase [NADP(+)]': 'O89000',\n",
       " 'Methylmalonate-semialdehyde dehydrogenase [acylating], mitochondrial': 'Q02253',\n",
       " 'Aldehyde dehydrogenase, mitochondrial': 'Q91ZD7',\n",
       " 'Membrane primary amine oxidase': 'Q5R1T5',\n",
       " 'Alpha-aminoadipic semialdehyde dehydrogenase': 'Q64057',\n",
       " 'Betaine--homocysteine S-methyltransferase 1': 'Q6AZ06',\n",
       " 'Methionine synthase': 'Q9I2Q2',\n",
       " 'Choline dehydrogenase, mitochondrial': 'Q6UPE0',\n",
       " 'Phosphatidylethanolamine N-methyltransferase': 'Q08388',\n",
       " 'Biotin--protein ligase': 'Q920N2',\n",
       " 'Biotinidase': 'Q5FVF9',\n",
       " 'Signal peptidase complex subunit 1': 'Q9D958',\n",
       " 'Enoyl-CoA hydratase, mitochondrial': 'P14604',\n",
       " 'Hydroxyacyl-coenzyme A dehydrogenase, mitochondrial': 'Q9WVK7',\n",
       " 'Acetyl-CoA acetyltransferase, mitochondrial': 'P24752',\n",
       " 'Hydroxymethylglutaryl-CoA synthase, mitochondrial': 'P22791',\n",
       " 'Succinyl-CoA:3-ketoacid coenzyme A transferase 1, mitochondrial': 'B2GV06',\n",
       " 'Hydroxymethylglutaryl-CoA lyase, mitochondrial': 'P97519',\n",
       " 'Acyl-coenzyme A synthetase ACSM1, mitochondrial': 'Q9XT43',\n",
       " 'Short-chain specific acyl-CoA dehydrogenase, mitochondrial': 'P15651',\n",
       " 'Cytochrome P450 1A2': 'Q64588',\n",
       " 'Cytochrome P450 2E1': 'P05182',\n",
       " 'Arylamine N-acetyltransferase 2': 'A4Z6T9',\n",
       " 'Cytochrome P450 3A4': 'P08684',\n",
       " 'Cytochrome P450 2C8': 'P10632',\n",
       " 'Cytochrome P450 2C9': 'P11712',\n",
       " 'Xanthine dehydrogenase/oxidase': 'Q63157',\n",
       " 'Cytochrome P450 2A6': 'P11509',\n",
       " 'Histone-lysine N-methyltransferase SETD7': 'D4ADE5',\n",
       " 'Trimethyllysine dioxygenase, mitochondrial': 'Q5U2P7',\n",
       " 'Serine hydroxymethyltransferase, cytosolic': 'Q8R0X9',\n",
       " '4-trimethylaminobutyraldehyde dehydrogenase': 'Q6GMM4',\n",
       " 'Gamma-butyrobetaine dioxygenase': 'Q9QZU7',\n",
       " 'Tyrosine 3-monooxygenase': 'P24529',\n",
       " 'Aromatic-L-amino-acid decarboxylase': 'Q6LEG4',\n",
       " 'Dopamine beta-hydroxylase': 'Q05754',\n",
       " 'Phenylethanolamine N-methyltransferase': 'P10937',\n",
       " 'Cysteine dioxygenase type 1': 'Q6NS32',\n",
       " '3-mercaptopyruvate sulfurtransferase': 'P97532',\n",
       " 'L-lactate dehydrogenase A-like 6A': 'Q6ZMR3',\n",
       " 'Cystinosin': 'P57757',\n",
       " 'Cysteine--tRNA ligase, cytoplasmic': 'Q9ER72',\n",
       " 'Extracellular superoxide dismutase [Cu-Zn]': 'Q64667',\n",
       " 'Superoxide dismutase [Cu-Zn]': 'Q9N5Y1',\n",
       " 'Superoxide dismutase [Mn], mitochondrial': 'Q9V7T8',\n",
       " 'Catalase': 'Q9VVT1',\n",
       " '5,6-dihydroxyindole-2-carboxylic acid oxidase': 'Q8WN57',\n",
       " 'Alcohol dehydrogenase 1B': 'P00325',\n",
       " 'Acetyl-coenzyme A synthetase, cytoplasmic': 'Q8BK97',\n",
       " 'Acetyl-coenzyme A synthetase 2-like, mitochondrial': 'Q8K0M6',\n",
       " 'Aldehyde dehydrogenase X, mitochondrial': 'Q66HF8',\n",
       " '3-hydroxyacyl-CoA dehydrogenase type-2': 'Q8MRC1',\n",
       " 'Trifunctional enzyme subunit alpha, mitochondrial': 'Q5BIZ5',\n",
       " 'Trifunctional enzyme subunit beta, mitochondrial': 'Q60587',\n",
       " 'Trans-2-enoyl-CoA reductase, mitochondrial': 'Q9BV79',\n",
       " '3-ketoacyl-CoA thiolase, mitochondrial': 'P13437',\n",
       " 'Palmitoyl-protein thioesterase 1': 'Q9W3C7',\n",
       " 'Carnitine O-palmitoyltransferase 2, mitochondrial': 'P97781',\n",
       " 'Carnitine O-palmitoyltransferase 1, liver isoform': 'Q6IMZ4',\n",
       " 'Long-chain-fatty-acid--CoA ligase 1': 'P18163',\n",
       " 'Long-chain specific acyl-CoA dehydrogenase, mitochondrial': 'P15650',\n",
       " 'Medium-chain specific acyl-CoA dehydrogenase, mitochondrial': 'P08503',\n",
       " 'Short/branched chain specific acyl-CoA dehydrogenase, mitochondrial': 'P70584',\n",
       " 'Glutaryl-CoA dehydrogenase, mitochondrial': 'Q60759',\n",
       " 'Very long-chain specific acyl-CoA dehydrogenase, mitochondrial': 'P45953',\n",
       " 'Dihydrofolate reductase': 'A0A0G5J004',\n",
       " 'C-1-tetrahydrofolate synthase, cytoplasmic': 'Q62808',\n",
       " 'Cytosolic 10-formyltetrahydrofolate dehydrogenase': 'Q5HZB2',\n",
       " 'Formimidoyltransferase-cyclodeaminase': 'Q5BKB7',\n",
       " '5-formyltetrahydrofolate cyclo-ligase': 'Q5M9F6',\n",
       " 'Methylenetetrahydrofolate reductase': 'D4A7E8',\n",
       " 'Folylpolyglutamate synthase, mitochondrial': 'Q99K38',\n",
       " 'Gamma-glutamyl hydrolase': 'Q62867',\n",
       " 'Monofunctional C1-tetrahydrofolate synthase, mitochondrial': 'Q8K2N5',\n",
       " 'Bifunctional methylenetetrahydrofolate dehydrogenase/cyclohydrolase, mitochondrial': 'Q3TMN4',\n",
       " 'Methionyl-tRNA formyltransferase, mitochondrial': 'Q5I0C5',\n",
       " 'Proton-coupled folate transporter': 'Q5EBA8',\n",
       " 'Aldose reductase': 'P07943',\n",
       " 'Sorbitol dehydrogenase': 'Q5I0F3',\n",
       " '14 kDa phosphohistidine phosphatase': 'Q9DAK9',\n",
       " 'Mannose-6-phosphate isomerase': 'Q68FX1',\n",
       " 'Phosphomannomutase 1': 'O35621',\n",
       " '6-phosphofructokinase, liver type': 'P17858',\n",
       " 'Fructose-1,6-bisphosphatase 1': 'Q64594',\n",
       " 'Mannose-1-phosphate guanyltransferase beta': 'Q8BTZ7',\n",
       " 'GDP-mannose 4,6 dehydratase': 'Q8K0C9',\n",
       " 'GDP-L-fucose synthase': 'P23591',\n",
       " 'Fucose-1-phosphate guanylyltransferase': 'A0A0G2K0C0',\n",
       " 'L-fucose kinase': 'Q7TMC8',\n",
       " 'Ketohexokinase': 'P97551',\n",
       " 'Triosephosphate isomerase': 'Q9HV51',\n",
       " 'Fructose-bisphosphate aldolase A': 'Q63038',\n",
       " '6-phosphofructo-2-kinase/fructose-2,6-bisphosphatase 1': 'P16119',\n",
       " 'Fructose-bisphosphate aldolase B': 'P70706',\n",
       " 'Glyceraldehyde-3-phosphate dehydrogenase': 'A0A0G5ZZ99',\n",
       " 'Glycerol-3-phosphate dehydrogenase [NAD(+)], cytoplasmic': 'P13707',\n",
       " 'Glycerol-3-phosphate dehydrogenase, mitochondrial': 'P35571',\n",
       " 'D-beta-hydroxybutyrate dehydrogenase, mitochondrial': 'Q5U2Q2',\n",
       " 'Alpha-aminoadipic semialdehyde synthase, mitochondrial': 'A2VCW9',\n",
       " 'Kynurenine/alpha-aminoadipate aminotransferase, mitochondrial': 'Q64602',\n",
       " 'Low affinity cationic amino acid transporter 2': 'P52569',\n",
       " 'Mitochondrial ornithine transporter 2': 'Q9BXI2',\n",
       " 'Peroxisomal sarcosine oxidase': 'Q9D826',\n",
       " 'Probable 2-oxoglutarate dehydrogenase E1 component DHKTD1, mitochondrial': 'Q96HY7',\n",
       " 'Calcium-binding mitochondrial carrier protein Aralar1': 'A2BFG0',\n",
       " 'Mitochondrial 2-oxoglutarate/malate carrier protein': 'P97700',\n",
       " 'Glucokinase': 'A6V2N1',\n",
       " 'Phosphoglucomutase-1': 'P38652',\n",
       " 'UTP--glucose-1-phosphate uridylyltransferase': 'Q59633',\n",
       " 'UDP-glucose 6-dehydrogenase': 'O70199',\n",
       " 'UDP-glucuronic acid decarboxylase 1': 'Q8K464',\n",
       " 'Galactose-1-phosphate uridylyltransferase': 'Q4KM61',\n",
       " 'Galactokinase': 'Q9JIA6',\n",
       " 'UDP-glucose 4-epimerase': 'P18645',\n",
       " 'Pantetheinase': 'A6QPH4',\n",
       " 'Pantothenate kinase 1': 'Q9QXM8',\n",
       " 'Phosphopantothenate--cysteine ligase': 'Q9DA06',\n",
       " 'Phosphopantothenoylcysteine decarboxylase': 'D3ZZZ5',\n",
       " 'Bifunctional coenzyme A synthase': 'Q8K3Y4',\n",
       " 'Ectonucleotide pyrophosphatase/phosphodiesterase family member 1': 'Q920C8',\n",
       " 'Glycine N-acyltransferase': 'Q7TP56',\n",
       " 'Acyl-coenzyme A synthetase ACSM2B, mitochondrial': 'Q68CK6',\n",
       " 'Lactoylglutathione lyase': 'Q9HU72',\n",
       " 'Probable D-lactate dehydrogenase, mitochondrial': 'Q8CIV4',\n",
       " 'Hydroxyacylglutathione hydrolase, mitochondrial': 'Q4V8M8',\n",
       " 'Tyrosinase': 'P14679',\n",
       " 'FAD synthase': 'D4A4P4',\n",
       " 'Low molecular weight phosphotyrosine protein phosphatase': 'Q9R138',\n",
       " 'Riboflavin kinase': 'Q8CFV9',\n",
       " 'Thiamin pyrophosphokinase 1': 'Q0P5L2',\n",
       " 'Thiamine transporter 1': 'Q8R4Y1',\n",
       " 'Cancer-related nucleoside-triphosphatase': 'Q9BSD7',\n",
       " 'Thiamine-triphosphatase': 'Q8CGV7',\n",
       " 'Spermidine synthase': 'Q99MI5',\n",
       " 'Spermine synthase': 'Q3MIE9',\n",
       " 'Ornithine decarboxylase': 'Q94278',\n",
       " 'S-adenosylmethionine decarboxylase proenzyme': 'A6UZF5',\n",
       " '2-aminoethanethiol dioxygenase': 'G3MZK0',\n",
       " 'Cysteine sulfinic acid decarboxylase': 'Q9DBE0',\n",
       " '4-hydroxybenzoate polyprenyltransferase, mitochondrial': 'Q5DX46',\n",
       " 'Hexaprenyldihydroxybenzoate methyltransferase, mitochondrial': 'Q9NZJ6',\n",
       " 'Ubiquinone biosynthesis monooxygenase COQ6': 'Q9Y2Z9',\n",
       " '2-methoxy-6-polyprenyl-1,4-benzoquinol methylase, mitochondrial': 'P34666',\n",
       " 'Ubiquinone biosynthesis protein COQ7 homolog': 'Q99807',\n",
       " 'Sulfite oxidase, mitochondrial': 'Q6P6W0',\n",
       " \"Bifunctional 3'-phosphoadenosine 5'-phosphosulfate synthase 2\": 'Q9Z274',\n",
       " \"3'(2'),5'-bisphosphate nucleotidase 1\": 'Q9Z1N4',\n",
       " 'Estrogen sulfotransferase': 'P19217',\n",
       " 'Sulfotransferase 1A1': 'Q548D2',\n",
       " 'Sulfotransferase family cytosolic 2B member 1': 'Q29YR6',\n",
       " 'Carbohydrate sulfotransferase 11': 'P69478',\n",
       " 'Very long-chain acyl-CoA synthetase': 'P97524',\n",
       " 'Phytanoyl-CoA dioxygenase, peroxisomal': 'Q9QY64',\n",
       " '2-hydroxyacyl-CoA lyase 1': 'Q8CHM7',\n",
       " 'Fatty aldehyde dehydrogenase': 'P30839',\n",
       " 'ATP-binding cassette sub-family D member 1': 'Q9QZ32',\n",
       " 'ATP-binding cassette sub-family D member 2': 'Q9QY44',\n",
       " 'Tyrosine aminotransferase': 'Q5EBB6',\n",
       " 'Phenylalanine-4-hydroxylase': 'P04176',\n",
       " 'Tyrosine--tRNA ligase, cytoplasmic': 'Q4KM49',\n",
       " 'Phenylalanine--tRNA ligase alpha subunit': 'Q02NN7',\n",
       " 'Phenylalanine--tRNA ligase beta subunit': 'Q9I0A4',\n",
       " '4-hydroxyphenylpyruvate dioxygenase': 'O88655',\n",
       " 'Homogentisate 1,2-dioxygenase': 'Q7TPP2',\n",
       " 'Maleylacetoacetate isomerase': 'B0BNI1',\n",
       " 'Fumarylacetoacetase': 'P25093',\n",
       " 'Carnosine synthase 1': 'D3Z945',\n",
       " 'Protein arginine N-methyltransferase 3': 'O70467',\n",
       " 'Histidine decarboxylase': 'P16453',\n",
       " 'Histamine N-methyltransferase': 'Q59JM9',\n",
       " 'Amine oxidase [flavin-containing] A': 'Q63817',\n",
       " 'Aldehyde dehydrogenase, dimeric NADP-preferring': 'P11883',\n",
       " 'Urocanate hydratase': 'Q8VC12',\n",
       " 'Probable imidazolonepropionase': 'Q9DBA8',\n",
       " 'Amiloride-sensitive amine oxidase [copper-containing]': 'Q63973',\n",
       " 'Carnosine N-methyltransferase': 'Q80UY1',\n",
       " 'cytosolic non-specific dipeptidase': 'Q96KP4',\n",
       " 'Probable histidine--tRNA ligase, mitochondrial': 'A5D7V9',\n",
       " 'Histidine--tRNA ligase, cytoplasmic': 'Q2KI84',\n",
       " 'Leukotriene-B(4) omega-hydroxylase 1': 'P78329',\n",
       " 'Leukotriene A-4 hydrolase': 'P30349',\n",
       " 'Leukotriene C4 synthase': 'Q925U2',\n",
       " 'Prostacyclin synthase': 'Q62969',\n",
       " 'Prostaglandin-H2 D-isomerase': 'P22057',\n",
       " 'Aldo-keto reductase family 1 member C3': 'P42330',\n",
       " 'Thromboxane-A synthase': 'P49430',\n",
       " 'Prostaglandin E synthase': 'Q3T9C5',\n",
       " 'Prostamide/prostaglandin F synthase': 'D3ZVR7',\n",
       " 'Carbonyl reductase [NADPH] 1': 'Q3KR58',\n",
       " 'Arachidonate 15-lipoxygenase': 'F1MA51',\n",
       " 'Cytochrome P450 2B6': 'P20813',\n",
       " 'Bifunctional epoxide hydrolase 2': 'P80299',\n",
       " 'Arachidonate 5-lipoxygenase': 'P12527',\n",
       " 'Prostaglandin G/H synthase 1': 'Q63684',\n",
       " 'Arachidonate 15-lipoxygenase B': 'Q8K4F2',\n",
       " 'Arachidonate 12-lipoxygenase, 12R-type': 'Q2KMM5',\n",
       " 'Cytochrome P450 2J2': 'P51589',\n",
       " 'Cytochrome P450 2U1': 'Q4V8D1',\n",
       " 'Cytochrome P450 4A11': 'Q02928',\n",
       " 'Arachidonate 12-lipoxygenase, 12S-type': 'F1LQ70',\n",
       " 'Prostaglandin G/H synthase 2': 'Q925V4',\n",
       " 'Cytochrome P450 4F8': 'P98187',\n",
       " 'Leukotriene-B(4) omega-hydroxylase 2': 'Q3MID2',\n",
       " 'Cytochrome P450 19A1': 'P11511',\n",
       " 'Estradiol 17-beta-dehydrogenase 1': 'P51657',\n",
       " 'UDP-glucuronosyltransferase 2B11': 'O75310',\n",
       " 'Steroid 17-alpha-hydroxylase/17,20 lyase': 'Q6LAE5',\n",
       " '3 beta-hydroxysteroid dehydrogenase/Delta 5-->4-isomerase type 1': 'P14060',\n",
       " 'Testosterone 17-beta-dehydrogenase 3': 'O54939',\n",
       " '3-oxo-5-beta-steroid 4-dehydrogenase': 'P31210',\n",
       " '3-oxo-5-alpha-steroid 4-dehydrogenase 1': 'A0JPJ6',\n",
       " 'UDP-glucuronosyltransferase 2B17': 'P17717',\n",
       " 'Steryl-sulfatase': 'P15589',\n",
       " 'Alanine aminotransferase 2': 'Q8K1J3',\n",
       " 'Sodium-coupled neutral amino acid transporter 4': 'Q969I6',\n",
       " 'Solute carrier family 2, facilitated glucose transporter member 2': 'P11168',\n",
       " 'Solute carrier family 2, facilitated glucose transporter member 4': 'P14672',\n",
       " 'Mitochondrial glutamate carrier 1': 'Q08DK4',\n",
       " 'Neutral amino acid transporter A': 'O35874',\n",
       " 'NAD(P)H dehydrogenase [quinone] 1': 'Q63478',\n",
       " 'Vitamin K-dependent gamma-carboxylase': 'O88496',\n",
       " 'Vitamin K epoxide reductase complex subunit 1': 'Q6TEK4',\n",
       " '1-Acyl-sn-glycerol-3-phosphate acyltransferase alpha': 'Q99943',\n",
       " 'Phosphatidate cytidylyltransferase 1': 'O88208',\n",
       " 'CDP-diacylglycerol--inositol 3-phosphatidyltransferase': 'P70500',\n",
       " 'Acetylcholinesterase': 'Q9VFY0',\n",
       " 'Choline O-acetyltransferase': 'P32756',\n",
       " 'Glycerophosphodiester phosphodiesterase 1': 'Q6IRJ6',\n",
       " 'Acyl-protein thioesterase 1': 'P70470',\n",
       " 'Group XV phospholipase A2': 'Q675A5',\n",
       " 'Group IID secretory phospholipase A2': 'Q9JLK0',\n",
       " 'Phosphoethanolamine/phosphocholine phosphatase': 'Q9FZ62',\n",
       " 'Cholinephosphotransferase 1': 'Q66H21',\n",
       " 'Diacylglycerol kinase alpha': 'P51556',\n",
       " 'Lipid phosphate phosphohydrolase 1': 'O14494',\n",
       " 'Phospholipase D2': 'Q08DV7',\n",
       " 'Phosphatidylserine synthase 2': 'B2GV22',\n",
       " 'Phosphatidylserine decarboxylase proenzyme': 'B7V346',\n",
       " 'CDP-diacylglycerol--glycerol-3-phosphate 3-phosphatidyltransferase, mitochondrial': 'Q2KJ28',\n",
       " 'Phosphatidylglycerophosphatase and protein-tyrosine phosphatase 1': 'G3V7B6',\n",
       " 'Cardiolipin synthase': 'Q9UJA2',\n",
       " 'Glycerol-3-phosphate acyltransferase 1, mitochondrial': 'Q61586',\n",
       " 'Choline-phosphate cytidylyltransferase A': 'P19836',\n",
       " 'Choline kinase alpha': 'F1MI39',\n",
       " 'Phosphatidylserine synthase 1': 'A0A0G2JZJ5',\n",
       " 'Glucose-6-phosphatase': 'Q5FVC9',\n",
       " 'Nucleoside diphosphate kinase B': 'P19804',\n",
       " 'UMP-CMP kinase': 'Q4KM73',\n",
       " 'Solute carrier family 2, facilitated glucose transporter member 1': 'P11166',\n",
       " 'Alpha-lactalbumin': 'P00715',\n",
       " 'Beta-1,4-galactosyltransferase 1': 'P15535',\n",
       " 'Soluble calcium-activated nucleotidase 1': 'Q4V8N9',\n",
       " 'UDP-galactose translocator': 'Q9R0M8',\n",
       " 'Acetyl-CoA acetyltransferase, cytosolic': 'Q5XI22',\n",
       " 'Hydroxymethylglutaryl-CoA synthase, cytoplasmic': 'Q01581',\n",
       " 'Mevalonate kinase': 'P46086',\n",
       " 'Phosphomevalonate kinase': 'Q15126',\n",
       " 'Diphosphomevalonate decarboxylase': 'P53602',\n",
       " 'Isopentenyl-diphosphate Delta-isomerase 1': 'Q13907',\n",
       " 'Farnesyl pyrophosphate synthase': 'P14324',\n",
       " 'Geranylgeranyl pyrophosphate synthase': 'Q6F596',\n",
       " 'Squalene synthase': 'P37268',\n",
       " 'Squalene monooxygenase': 'Q14534',\n",
       " 'Lanosterol synthase': 'Q1G1A4',\n",
       " 'Delta(24)-sterol reductase': 'Q39085',\n",
       " 'Lanosterol 14-alpha demethylase': 'Q16850',\n",
       " 'Delta(14)-sterol reductase': 'Q9LDR4',\n",
       " 'Methylsterol monooxygenase 1': 'Q15800',\n",
       " 'Sterol-4-alpha-carboxylate 3-dehydrogenase, decarboxylating': 'Q15738',\n",
       " '3-keto-steroid reductase': 'P56937',\n",
       " '3-beta-hydroxysteroid-Delta(8),Delta(7)-isomerase': 'Q15125',\n",
       " 'Lathosterol oxidase': 'O75845',\n",
       " '3-hydroxy-3-methylglutaryl-coenzyme A reductase': 'P04035',\n",
       " 'Sterol O-acyltransferase 1': 'O70536',\n",
       " 'Lysosomal acid lipase/cholesteryl ester hydrolase': 'Q64194',\n",
       " '2-oxoisovalerate dehydrogenase subunit beta, mitochondrial': 'P35738',\n",
       " '2-oxoisovalerate dehydrogenase subunit alpha, mitochondrial': 'P11960',\n",
       " 'Branched-chain-amino-acid aminotransferase, cytosolic': 'P54690',\n",
       " 'Lipoamide acyltransferase component of branched-chain alpha-keto acid dehydrogenase complex, mitochondrial': 'Q3TMF5',\n",
       " 'Isovaleryl-CoA dehydrogenase, mitochondrial': 'P12007',\n",
       " 'Methylcrotonoyl-CoA carboxylase subunit alpha, mitochondrial': 'Q5I0C3',\n",
       " 'Methylcrotonoyl-CoA carboxylase beta chain, mitochondrial': 'Q5XIT9',\n",
       " 'Isobutyryl-CoA dehydrogenase, mitochondrial': 'Q9D7B6',\n",
       " '3-hydroxyisobutyryl-CoA hydrolase, mitochondrial': 'Q5XIE6',\n",
       " '3-hydroxyisobutyrate dehydrogenase, mitochondrial': 'A1L107',\n",
       " 'Propionyl-CoA carboxylase alpha chain, mitochondrial': 'P14882',\n",
       " 'Propionyl-CoA carboxylase beta chain, mitochondrial': 'P07633',\n",
       " 'Methylmalonyl-CoA epimerase, mitochondrial': 'Q9D1I5',\n",
       " 'Methylmalonyl-CoA mutase, mitochondrial': 'P16332',\n",
       " 'Aldehyde oxidase': 'Q06278',\n",
       " 'Methylglutaconyl-CoA hydratase, mitochondrial': 'F1LU71',\n",
       " \"Bis(5'-nucleosyl)-tetraphosphatase [asymmetrical]\": 'Q6PEC0',\n",
       " 'Nucleoside diphosphate kinase 6': 'O88426',\n",
       " 'Ectonucleoside triphosphate diphosphohydrolase 5': 'Q6P6S9',\n",
       " 'Ectonucleoside triphosphate diphosphohydrolase 8': 'Q5DRK1',\n",
       " 'Guanylate kinase': 'Q9HTM2',\n",
       " \"Cytosolic purine 5'-nucleotidase\": 'Q9D8G6',\n",
       " 'Deoxyguanosine kinase, mitochondrial': 'Q9WUT9',\n",
       " 'Adenine phosphoribosyltransferase': 'A0A071L007',\n",
       " 'Purine nucleoside phosphorylase': 'P00491',\n",
       " 'Hypoxanthine-guanine phosphoribosyltransferase': 'Q62926',\n",
       " 'Guanine deaminase': 'Q9VMY9',\n",
       " 'ADP-sugar pyrophosphatase': 'Q6AY63',\n",
       " 'Ribose-phosphate pyrophosphokinase 3': 'P38689',\n",
       " 'Trifunctional purine biosynthetic protein adenosine-3': 'G3V918',\n",
       " 'Phosphoribosylformylglycinamidine synthase': 'Q9HXN2',\n",
       " 'Multifunctional protein ADE2': 'P51583',\n",
       " 'Bifunctional purine biosynthesis protein PURH': 'Q6IN16',\n",
       " 'Inosine triphosphate pyrophosphatase': 'D3ZW55',\n",
       " \"Inosine-5'-monophosphate dehydrogenase 1\": 'D3ZLZ7',\n",
       " 'GMP reductase 1': 'Q9Z244',\n",
       " 'Adenylosuccinate synthetase isozyme 2': 'D4AEP0',\n",
       " 'AMP deaminase 1': 'Q6LDJ4',\n",
       " 'Adenosine deaminase': 'Q920P6',\n",
       " 'Adenylate kinase isoenzyme 1': 'Q9JJN3',\n",
       " 'ATPase family AAA domain-containing protein 1': 'B3STU2',\n",
       " \"cAMP-specific 3',5'-cyclic phosphodiesterase 4D\": 'Q8CG06',\n",
       " 'Adenylate cyclase type 10': 'Q96PN6',\n",
       " 'Guanylate cyclase soluble subunit alpha-2': 'Q54A43',\n",
       " 'Guanylate cyclase soluble subunit beta-1': 'Q8CH88',\n",
       " \"cAMP and cAMP-inhibited cGMP 3',5'-cyclic phosphodiesterase 10A\": 'Q9QYJ5',\n",
       " 'Thioredoxin': 'P11232',\n",
       " 'Ribonucleoside-diphosphate reductase large subunit': 'Q91YM8',\n",
       " 'Ribonucleoside-diphosphate reductase subunit M2': 'Q4KLN6',\n",
       " 'Alkaline phosphatase, tissue-nonspecific isozyme': 'P70707',\n",
       " \"Pyridoxine-5'-phosphate oxidase\": 'O88794',\n",
       " 'Pyridoxal kinase': 'O35331',\n",
       " 'Pyridoxal phosphate phosphatase': 'Q8VD52',\n",
       " 'Pyruvate kinase isozymes R/L': 'P30613',\n",
       " 'Phosphoenolpyruvate carboxykinase, cytosolic [GTP]': 'P07379',\n",
       " 'Malate dehydrogenase, cytoplasmic': 'Q6PCV2',\n",
       " 'NAD-dependent malic enzyme, mitochondrial': 'P23368',\n",
       " 'Glyoxylate reductase/hydroxypyruvate reductase': 'Q91Z53',\n",
       " 'NADP-dependent malic enzyme': 'P13697',\n",
       " 'Acetyl-CoA carboxylase 1': 'P97902',\n",
       " 'Acyl-coenzyme A thioesterase 12': 'Q99NB7',\n",
       " 'Acylphosphatase-1': 'Q6P8S7',\n",
       " 'Glucose-6-phosphate 1-dehydrogenase': 'P05370',\n",
       " '6-phosphogluconolactonase': 'P85971',\n",
       " 'Ribose-5-phosphate isomerase': 'P47968',\n",
       " 'Ribokinase': 'Q9I2F4',\n",
       " 'Ribulose-phosphate 3-epimerase': 'A0A0G2JW38',\n",
       " 'Transketolase': 'P50137',\n",
       " 'Transaldolase': 'Q6PCV1',\n",
       " 'Putative deoxyribose-phosphate aldolase': 'Q19264',\n",
       " 'Glucose-6-phosphate isomerase': 'A0A072ZX40',\n",
       " '6-phosphogluconate dehydrogenase, decarboxylating': 'Q9DCD0',\n",
       " 'DNA (cytosine-5)-methyltransferase 1': 'Q9WU57',\n",
       " \"S-methyl-5'-thioadenosine phosphorylase\": 'Q3TJS4',\n",
       " 'Methionine-R-sulfoxide reductase B3': 'Q8IXL7',\n",
       " 'Methionine-R-sulfoxide reductase B2, mitochondrial': 'Q4FZX5',\n",
       " 'Solute carrier family 22 member 6': 'Q4U2R8',\n",
       " 'Solute carrier family 22 member 8': 'Q8TCC7',\n",
       " 'Cytochrome P450 2C19': 'P33261',\n",
       " 'UDP-glucuronosyltransferase 2B7': 'P16662',\n",
       " 'UDP-glucuronosyltransferase 2B4': 'P06133',\n",
       " 'DnaJ homolog subfamily B member 11': 'Q6TUG0',\n",
       " 'HSP90B1 protein': 'Q96GW1',\n",
       " '78 kDa glucose-regulated protein': 'P11021',\n",
       " 'HYOU1 protein': 'Q6IN67',\n",
       " 'Protein disulfide-isomerase A2': 'Q14AV9',\n",
       " 'Protein disulfide-isomerase A4': 'Q6P7S5',\n",
       " 'Protein disulfide-isomerase A6': 'Q641Y3',\n",
       " 'Peptidyl-prolyl cis-trans isomerase B': 'O88541',\n",
       " 'Stromal cell-derived factor 2-like protein 1': 'Q9ESP2',\n",
       " 'UDP-glucuronosyltransferase 1-1': 'Q64635',\n",
       " 'Endoplasmic reticulum resident protein 29': 'Q5BKC2',\n",
       " 'UDP-glucuronosyltransferase 1-3': 'P35503',\n",
       " 'UDP-glucuronosyltransferase 1-9': 'O60656',\n",
       " 'UDP-glucuronosyltransferase 1-10': 'Q9HAW8',\n",
       " 'Alcohol dehydrogenase 1A': 'P07327',\n",
       " 'Cytochrome P450 2D6': 'P10635',\n",
       " 'GTP cyclohydrolase 1': 'Q9Y0C8',\n",
       " 'GTP cyclohydrolase 1 feedback regulatory protein': 'P70552',\n",
       " '6-pyruvoyl tetrahydrobiopterin synthase': 'Q9VIP5',\n",
       " 'Sepiapterin reductase': 'P18297',\n",
       " 'Dihydropteridine reductase': 'P11348',\n",
       " 'Aldo-keto reductase family 1 member C4': 'P17516',\n",
       " 'Cytochrome P450 11B2, mitochondrial': 'P19099',\n",
       " 'Cytochrome P450 11B1, mitochondrial': 'Q29457',\n",
       " 'Corticosteroid 11-beta-dehydrogenase isozyme 1': 'Q6LDH6',\n",
       " 'Steroid 21-hydroxylase': 'P08686',\n",
       " 'Cholesterol side-chain cleavage enzyme, mitochondrial': 'P05108',\n",
       " 'Corticosteroid 11-beta-dehydrogenase isozyme 2': 'P50233',\n",
       " '3 beta-hydroxysteroid dehydrogenase/Delta 5-->4-isomerase type 2': 'P26439',\n",
       " 'L-dopachrome tautomerase': 'P40126',\n",
       " 'Macrophage migration inhibitory factor': 'P30904',\n",
       " 'Catechol O-methyltransferase': 'P22734',\n",
       " '3-hydroxyanthranilate 3,4-dioxygenase': 'Q64556',\n",
       " 'Inositol-3-phosphate synthase 1': 'Q6AYK3',\n",
       " 'Inositol polyphosphate multikinase': 'Q99NI4',\n",
       " 'Inositol-tetrakisphosphate 1-kinase': 'D3ZQM7',\n",
       " 'Inositol-pentakisphosphate 2-kinase': 'Q9XXI1',\n",
       " 'Multiple inositol polyphosphate phosphatase 1': 'Q8T3I2',\n",
       " 'Inositol-trisphosphate 3-kinase A': 'P17105',\n",
       " 'Phosphatidylinositol 4,5-bisphosphate 5-phosphatase A': 'Q9JMC1',\n",
       " 'Inositol polyphosphate 1-phosphatase': 'Q9VFP6',\n",
       " 'Inositol monophosphatase 1': 'P97697',\n",
       " 'Inositol oxygenase': 'Q99PN7',\n",
       " 'Type I inositol 3,4-bisphosphate 4-phosphatase': 'Q62784',\n",
       " '1-phosphatidylinositol 4,5-bisphosphate phosphodiesterase delta-3': 'Q8BL19',\n",
       " '1-phosphatidylinositol-3-phosphate 5-kinase': 'Q9Y2I7',\n",
       " 'Polyphosphoinositide phosphatase': 'Q6A092',\n",
       " 'Protein VAC14 homolog': 'Q80W92',\n",
       " 'Phosphatidylinositol 4-kinase alpha': 'A0A0G2K2J3',\n",
       " 'Phosphatidylinositol 3-kinase catalytic subunit type 3': 'O88763',\n",
       " 'Phosphoinositide 3-kinase regulatory subunit 4': 'P0C0R5',\n",
       " 'Beclin-1': 'Q91XJ1',\n",
       " 'Activating molecule in BECN1-regulated autophagy protein 1': 'Q8BYW8',\n",
       " 'Phosphatidylinositol-4,5-bisphosphate 3-kinase catalytic subunit alpha isoform': 'P42336',\n",
       " 'Glycerate kinase': 'A0A140SB49',\n",
       " 'Glycerol kinase': 'Q63060',\n",
       " 'Lipid phosphate phosphohydrolase 2': 'O43688',\n",
       " 'Lipoprotein lipase': 'Q06000',\n",
       " 'Hepatic triacylglycerol lipase': 'P07867',\n",
       " 'Sterol 26-hydroxylase, mitochondrial': 'Q64639',\n",
       " 'Bile acyl-CoA synthetase': 'Q9ES38',\n",
       " 'Alpha-methylacyl-CoA racemase': 'O09176',\n",
       " 'Peroxisomal acyl-coenzyme A oxidase 2': 'P97562',\n",
       " 'Peroxisomal multifunctional enzyme type 2': 'P70540',\n",
       " 'Non-specific lipid-transfer protein': 'Q642G0',\n",
       " 'Bile acid-CoA:amino acid N-acyltransferase': 'Q5M8A2',\n",
       " '3 beta-hydroxysteroid dehydrogenase type 7': 'O35048',\n",
       " '7-alpha-hydroxycholest-4-en-3-one 12-alpha-hydroxylase': 'O88962',\n",
       " 'Cholesterol 7-alpha-monooxygenase': 'P51543',\n",
       " 'Cholesterol 24-hydroxylase': 'Q9WVK8',\n",
       " '24-hydroxycholesterol 7-alpha-hydroxylase': 'Q9JKJ9',\n",
       " '25-hydroxycholesterol 7-alpha-hydroxylase': 'Q63688',\n",
       " 'Cholesterol 25-hydroxylase': 'Q4QQV7',\n",
       " 'Hexokinase-2': 'Q9QWR1',\n",
       " 'Aldose 1-epimerase': 'Q66HG4',\n",
       " '6-phosphofructokinase, muscle type': 'P08237',\n",
       " 'Bisphosphoglycerate mutase': 'Q543Z6',\n",
       " 'Phosphoglycerate mutase 1': 'Q6P0K6',\n",
       " 'Phosphoglycerate mutase 2': 'P16290',\n",
       " 'Alpha-enolase': 'Q6P504',\n",
       " 'Serine palmitoyltransferase 2': 'P97356',\n",
       " 'Serine palmitoyltransferase 1': 'Q8BH11',\n",
       " '3-ketodihydrosphingosine reductase': 'Q6GV12',\n",
       " 'Sphingosine kinase 2': 'Q9DBH6',\n",
       " 'Sphingosine-1-phosphate phosphatase 2': 'B2RWL5',\n",
       " 'Sphingosine-1-phosphate lyase 1': 'Q8CHN6',\n",
       " 'Sphingolipid delta(4)-desaturase/C4-hydroxylase DES2': 'Q6QHC5',\n",
       " 'Alkaline ceramidase 1': 'A1L3S2',\n",
       " 'Ceramide kinase': 'Q52KP2',\n",
       " 'Phosphatidylcholine:ceramide cholinephosphotransferase 1': 'Q7TSX5',\n",
       " 'Ectonucleotide pyrophosphatase/phosphodiesterase family member 7': 'Q5EZ72',\n",
       " '2-hydroxyacylsphingosine 1-beta-galactosyltransferase': 'Q09426',\n",
       " 'Galactocerebrosidase': 'Q3U3H7',\n",
       " 'Glucosylceramidase': 'B2RYC9',\n",
       " 'Ceramide glucosyltransferase': 'O55149',\n",
       " 'Beta-galactosidase': 'P23780',\n",
       " 'Beta-1,4-galactosyltransferase 6': 'O88419',\n",
       " 'Galactosylceramide sulfotransferase': 'Q9D8V6',\n",
       " 'Arylsulfatase A': 'Q9DC66',\n",
       " 'Alpha-galactosidase A': 'P51569',\n",
       " 'Sialidase-3': 'Q497C0',\n",
       " 'Alkaline ceramidase 3': 'F1M6P3',\n",
       " 'Malonyl-CoA decarboxylase, mitochondrial': 'Q9WUY2',\n",
       " 'Acyl-CoA synthetase short-chain family member 3, mitochondrial': 'Q14DH7',\n",
       " 'L-lactate dehydrogenase A-like 6B': 'Q3T056',\n",
       " 'Sucrase-isomaltase, intestinal': 'P23739',\n",
       " 'Glucose 1,6-bisphosphate synthase': 'Q8CAA7',\n",
       " 'Glycogen [starch] synthase, liver': 'Q99MF8',\n",
       " '1,4-alpha-glucan-branching enzyme': 'A0A0G2JTB2',\n",
       " 'Maltase-glucoamylase, intestinal': 'O43451',\n",
       " 'Alpha-amylase 1': 'P00687',\n",
       " 'Glycogen debranching enzyme': 'A0A0G6CHM8',\n",
       " 'Glycogen phosphorylase, liver form': 'P09811',\n",
       " 'Beta-glucuronidase': 'P06760',\n",
       " 'Nicotinamide N-methyltransferase': 'P34254',\n",
       " 'Nicotinamide phosphoribosyltransferase': 'Q2VT47',\n",
       " 'ADP-ribosyl cyclase 2': 'Q10588',\n",
       " 'Peroxisomal NADH pyrophosphatase NUDT12': 'Q29RH3',\n",
       " 'Nicotinamide mononucleotide adenylyltransferase 2': 'Q9BZQ4',\n",
       " 'NAD(P) transhydrogenase, mitochondrial': 'A4FUB4',\n",
       " 'NAD kinase': 'Q9HZC0',\n",
       " 'Nicotinamide riboside kinase 1': 'Q6AY91',\n",
       " 'Nicotinate-nucleotide pyrophosphorylase [carboxylating]': 'P30819',\n",
       " 'Glucose-6-phosphate translocase': 'O43826',\n",
       " 'Lactase-phlorizin hydrolase': 'Q63719',\n",
       " 'Sodium/glucose cotransporter 1': 'P97787',\n",
       " 'Sodium/potassium-transporting ATPase subunit alpha-1': 'P05023',\n",
       " 'Sodium/potassium-transporting ATPase subunit alpha-2': 'P50993',\n",
       " 'Sodium/potassium-transporting ATPase subunit alpha-3': 'P13637',\n",
       " 'Sodium/potassium-transporting ATPase subunit alpha-4': 'Q13733',\n",
       " 'Sodium/potassium-transporting ATPase subunit beta-1': 'P05026',\n",
       " 'Sodium/potassium-transporting ATPase subunit beta-2': 'P14415',\n",
       " 'Sodium/potassium-transporting ATPase subunit beta-3': 'P54709',\n",
       " 'ATPase, (Na+)/K+ transporting, beta 4 polypeptide': 'B7ZKV8',\n",
       " 'Sodium/potassium-transporting ATPase subunit gamma': 'P54710',\n",
       " 'ATP-citrate synthase': 'Q8VIQ1',\n",
       " 'Mitochondrial citrate transport protein': 'Q6LAP8',\n",
       " 'Peroxisomal membrane protein PEX13': 'Q9D0K1',\n",
       " 'Peroxisomal membrane protein PEX14': 'Q9Z2Z4',\n",
       " 'Carnitine O-acetyltransferase': 'Q704S8',\n",
       " 'Peroxisomal carnitine O-octanoyltransferase': 'P48033',\n",
       " 'Mitochondrial carnitine/acylcarnitine carrier protein': 'P97521',\n",
       " 'Diphosphoinositol polyphosphate phosphohydrolase 1': 'Q566C7',\n",
       " 'Inositol hexakisphosphate and diphosphoinositol-pentakisphosphate kinase 1': 'P0C644',\n",
       " 'Inositol hexakisphosphate kinase 1': 'Q9ESM0',\n",
       " 'Type II inositol 3,4-bisphosphate 4-phosphatase': 'O35825',\n",
       " '2-amino-3-ketobutyrate coenzyme A ligase, mitochondrial': 'O88986',\n",
       " 'Dimethylglycine dehydrogenase, mitochondrial': 'Q63342',\n",
       " 'Sarcosine dehydrogenase, mitochondrial': 'O88499',\n",
       " 'Serine hydroxymethyltransferase, mitochondrial': 'Q99K87',\n",
       " '5-aminolevulinate synthase, nonspecific, mitochondrial': 'Q6P782',\n",
       " 'Glycine N-methyltransferase': 'P13255',\n",
       " 'D-3-phosphoglycerate dehydrogenase': 'A0A0G5WCU2',\n",
       " 'Phosphoserine aminotransferase': 'Q02PX3',\n",
       " 'Phosphoserine phosphatase': 'Q5M819',\n",
       " 'Serine racemase': 'Q76EQ0',\n",
       " 'Serine--tRNA ligase, cytoplasmic': 'Q6P799',\n",
       " 'Glycine--tRNA ligase': 'Q5I0G4',\n",
       " 'Delta-aminolevulinic acid dehydratase': 'Q59643',\n",
       " 'Porphobilinogen deaminase': 'A6VE37',\n",
       " 'Uroporphyrinogen-III synthase': 'P48246',\n",
       " 'Uroporphyrinogen decarboxylase': 'B7V3C7',\n",
       " 'Coproporphyrinogen-III oxidase, mitochondrial': 'P36551',\n",
       " 'Protoporphyrinogen oxidase': 'D3ZVN7',\n",
       " 'Ferritin, mitochondrial': 'Q9D5F4',\n",
       " 'Ferrochelatase, mitochondrial': 'P22315',\n",
       " 'Protoheme IX farnesyltransferase, mitochondrial': 'F1LVF4',\n",
       " 'Cytochrome c oxidase assembly protein COX15 homolog': 'Q921K5',\n",
       " 'Heme oxygenase 1': 'Q5BK87',\n",
       " 'Biliverdin reductase A': 'P46844',\n",
       " 'Feline leukemia virus subgroup C receptor-related protein 2': 'Q9UPI3',\n",
       " 'Lysosomal alpha-glucosidase': 'Q6P7A9',\n",
       " 'Dihydroorotate dehydrogenase (quinone), mitochondrial': 'Q63707',\n",
       " 'Uridine monophosphate synthetase isoform A': 'A8K5J1 ',\n",
       " 'UMP-CMP kinase 2, mitochondrial': 'Q9DC34',\n",
       " 'GTP:AMP phosphotransferase, mitochondrial': 'Q9VGU6',\n",
       " 'CTP synthase 1': 'Q922Y4',\n",
       " 'Uridine phosphorylase 2': 'Q8CGR7',\n",
       " 'Ribonucleoside-diphosphate reductase subunit M2 B': 'Q6PEE3',\n",
       " 'Uridine-cytidine kinase-like 1': 'Q91YL3',\n",
       " 'Cytidine deaminase': 'D4AC20',\n",
       " 'Thymidine phosphorylase': 'Q5FVR2',\n",
       " 'Deoxycytidylate deaminase': 'Q5M9G0',\n",
       " 'Thymidine kinase, cytosolic': 'Q6LAG4',\n",
       " 'Thymidylate synthase': 'Q9I6F1',\n",
       " \"Deoxyuridine 5'-triphosphate nucleotidohydrolase, mitochondrial\": 'P33316',\n",
       " 'Peroxisomal membrane protein 11C': 'Q6P6M5',\n",
       " 'Neutral amino acid transporter B(0)': 'Q61326',\n",
       " 'Tryptophan--tRNA ligase, mitochondrial': 'Q3T099',\n",
       " 'Tryptophan 2,3-dioxygenase': 'Q8I7L6',\n",
       " 'Kynureninase': 'Q18026',\n",
       " 'Kynurenine formamidase': 'M0RC77',\n",
       " 'Kynurenine 3-monooxygenase': 'Q21795',\n",
       " '2-amino-3-carboxymuconate-6-semialdehyde decarboxylase': 'Q8TDX5',\n",
       " 'Tryptophan 5-hydroxylase 1': 'P09810',\n",
       " 'Acetylserotonin O-methyltransferase': 'O09179',\n",
       " 'Indoleamine 2,3-dioxygenase 1': 'Q9ERD9',\n",
       " 'Serotonin N-acetyltransferase': 'Q64553',\n",
       " 'Cytochrome P450 1A1': 'P00185',\n",
       " 'Indolethylamine N-methyltransferase': 'Q9CZ50',\n",
       " 'Tryptophan--tRNA ligase, cytoplasmic': 'P17248',\n",
       " 'Retinol dehydrogenase 12': 'Q9D1Y4',\n",
       " 'Dehydrogenase/reductase SDR family member 4': 'Q6IRD4',\n",
       " 'Retinol dehydrogenase 16': 'O75452',\n",
       " 'Retinol dehydrogenase 8': 'Q9N126',\n",
       " 'Short-chain dehydrogenase/reductase 3': 'Q922A6',\n",
       " 'Dehydrogenase/reductase SDR family member 9': 'Q8VD48',\n",
       " 'Retinol dehydrogenase 11': 'Q9D0U5',\n",
       " 'Lecithin retinol acyltransferase': 'Q9JI61',\n",
       " 'Acyl-CoA wax alcohol acyltransferase 1': 'B2RW94',\n",
       " 'Diacylglycerol O-acyltransferase 1': 'Q9D7Q5',\n",
       " 'Retinoid isomerohydrolase': 'F1M8Q2',\n",
       " 'Patatin-like phospholipase domain-containing protein 4': 'P41247',\n",
       " \"Beta,beta-carotene 15,15'-monooxygenase\": 'Q9HAY6',\n",
       " 'Retinal dehydrogenase 1': 'O09184',\n",
       " 'Retinal dehydrogenase 2': 'Q4FZY8',\n",
       " 'Cytochrome P450 26A1': 'Q9R1F4',\n",
       " 'All-trans-retinol 13,14-reductase': 'Q8VHE9',\n",
       " 'Cytochrome P450 2A7': 'P20853',\n",
       " 'Cytochrome P450 2A13': 'Q16696',\n",
       " 'Cytochrome P450 3A7': 'P24462',\n",
       " 'Cytochrome P450 3A5': 'P20815',\n",
       " 'Cytochrome P450 3A43': 'Q9HB55',\n",
       " 'ADP/ATP translocase 1': 'Q05962',\n",
       " 'ATP synthase lipid-binding protein, mitochondrial': 'Q06055',\n",
       " 'ATP synthase subunit alpha, mitochondrial': 'P25705',\n",
       " 'ATP synthase subunit epsilon, mitochondrial': 'P56381',\n",
       " 'ATP synthase subunit beta, mitochondrial': 'P00830',\n",
       " 'ATP synthase subunit b, mitochondrial': 'P24539',\n",
       " 'ATP synthase subunit delta, mitochondrial': 'P30049',\n",
       " 'ATP synthase subunit gamma, mitochondrial': 'P36542',\n",
       " 'ATP synthase subunit a': 'Q9HT14',\n",
       " 'NADH dehydrogenase [ubiquinone] 1 alpha subcomplex subunit 1': 'Q9C9Z5',\n",
       " 'Cytochrome c': 'Q2KJD4',\n",
       " 'Cytochrome b-c1 complex subunit 1, mitochondrial': 'P98080',\n",
       " 'Cytochrome c oxidase subunit 1': 'P00395',\n",
       " 'Fatty acid synthase': 'Q64717',\n",
       " 'Synaptojanin-1': 'Q810Z8',\n",
       " 'Phosphatidylinositol 3-kinase regulatory subunit alpha': 'Q63790',\n",
       " 'Phosphatidylinositol 4-phosphate 5-kinase type-1 alpha': 'D3ZSI8',\n",
       " 'Phosphatidylinositol-5-phosphate 4-kinase type-2 alpha': 'P48426',\n",
       " 'Phosphatidylinositol 3,4,5-trisphosphate 5-phosphatase 1': 'P97573',\n",
       " '72 kDa inositol polyphosphate 5-phosphatase': 'Q9WVR1',\n",
       " 'Phosphatidylinositol-4-phosphate 3-kinase C2 domain-containing subunit alpha': 'O00443',\n",
       " 'Receptor tyrosine-protein kinase erbB-2': 'Q6P732',\n",
       " 'EGFR protein': 'Q504U8',\n",
       " 'Phosphatidylinositol-4,5-bisphosphate 3-kinase catalytic subunit delta isoform': 'O00329',\n",
       " 'Phosphatidylinositol 3,4,5-trisphosphate 3-phosphatase and dual-specificity protein phosphatase PTEN': 'Q542G1',\n",
       " '1-phosphatidylinositol 4,5-bisphosphate phosphodiesterase beta-1': 'Q9NQ66',\n",
       " 'Trehalase': 'Q9W2M3',\n",
       " 'Dihydroxyacetone phosphate acyltransferase': 'A4IF87',\n",
       " 'Alkyldihydroxyacetonephosphate synthase, peroxisomal': 'Q9EQR2',\n",
       " 'Choline/ethanolaminephosphotransferase 1': 'O82567',\n",
       " 'Solute carrier family 12 member 3': 'P55017',\n",
       " 'Chloride channel protein ClC-Ka': 'P51800',\n",
       " 'Solute carrier family 12 member 1': 'Q13621',\n",
       " 'Solute carrier family 12 member 6': 'Q9UHW9',\n",
       " 'Solute carrier family 22 member 2': 'O15244',\n",
       " 'Amiloride-sensitive sodium channel subunit alpha': 'P37088',\n",
       " 'Amiloride-sensitive sodium channel subunit beta': 'P51168',\n",
       " 'Amiloride-sensitive sodium channel subunit delta': 'P51172',\n",
       " 'Amiloride-sensitive sodium channel subunit gamma': 'P51170',\n",
       " 'Excitatory amino acid transporter 3': 'P43005',\n",
       " 'Large neutral amino acids transporter small subunit 1': 'Q01650',\n",
       " '4F2 cell-surface antigen heavy chain': 'P08195',\n",
       " 'Sodium- and chloride-dependent transporter XTRP3': 'Q9NP91',\n",
       " 'Y+L amino acid transporter 1': 'Q9UM01',\n",
       " 'Y+L amino acid transporter 2': 'Q92536',\n",
       " 'Mineralocorticoid receptor': 'Q64174',\n",
       " 'Neutral and basic amino acid transport protein rBAT': 'Q07837',\n",
       " 'B(0,+)-type amino acid transporter 1': 'P82251',\n",
       " 'Urea transporter 2': 'Q15849',\n",
       " 'Large neutral amino acids transporter small subunit 2': 'Q9UHI5',\n",
       " 'Neuronal acetylcholine receptor subunit beta-2': 'P17787',\n",
       " 'Neuronal acetylcholine receptor subunit alpha-4': 'P43681',\n",
       " 'Voltage-dependent P/Q-type calcium channel subunit alpha-1A': 'O00555',\n",
       " 'Voltage-dependent L-type calcium channel subunit beta-1': 'Q02641',\n",
       " 'Voltage-dependent calcium channel subunit alpha-2/delta-2': 'Q9NY47',\n",
       " 'Sodium channel protein type 10 subunit alpha': 'Q9Y5Y9',\n",
       " 'Sodium channel subunit beta-1': 'Q07699',\n",
       " 'Potassium voltage-gated channel subfamily D member 2': 'Q9NZV8',\n",
       " 'Kv channel-interacting protein 2': 'Q9JM60',\n",
       " 'Pro-opiomelanocortin': 'P01194',\n",
       " 'Beta-endorphin': 'P01203',\n",
       " 'Neuroendocrine convertase 2': 'P16519',\n",
       " 'Glutamate [NMDA] receptor subunit zeta-1': 'Q05586',\n",
       " 'Glutamate [NMDA] receptor subunit epsilon-1': 'Q12879',\n",
       " 'Glutamate [NMDA] receptor subunit 3A': 'Q8TCU5',\n",
       " 'Alpha-1A adrenergic receptor': 'P18130',\n",
       " '5-hydroxytryptamine receptor 1A': 'P08908',\n",
       " 'Sodium-dependent noradrenaline transporter': 'P23975',\n",
       " 'Sodium-dependent serotonin transporter': 'P31645',\n",
       " 'Sodium-dependent dopamine transporter': 'Q01959',\n",
       " 'D(1A) dopamine receptor': 'Q8WND7',\n",
       " 'Mu-type opioid receptor': 'P35372',\n",
       " 'Angiotensin-converting enzyme': 'Q9EQM9',\n",
       " 'Renin': 'P00797',\n",
       " 'Angiotensinogen': 'P01015',\n",
       " 'Bevacizumab': nan,\n",
       " 'Vascular endothelial growth factor A': 'P15692',\n",
       " 'Vascular endothelial growth factor receptor 2': 'P35968',\n",
       " 'Vascular endothelial growth factor receptor 1': 'P17948',\n",
       " 'DNA topoisomerase 1': 'P11387',\n",
       " 'Multidrug resistance protein 1': 'P08183',\n",
       " 'Solute carrier organic anion transporter family member 1B1': 'Q9Y6L6',\n",
       " 'Multidrug resistance-associated protein 1': 'Q8HXQ5',\n",
       " 'Canalicular multispecific organic anion transporter 1': 'Q92887',\n",
       " 'ATP-binding cassette sub-family G member 2': 'Q9UNQ0',\n",
       " 'Liver carboxylesterase 1': 'P23141',\n",
       " 'Cocaine esterase': 'O00748',\n",
       " 'Cholinesterase': 'P06276',\n",
       " 'Tubulin beta-1 chain': 'Q9H4B7',\n",
       " 'Tubulin alpha-1B chain': 'P68363',\n",
       " 'Solute carrier organic anion transporter family member 1B3': 'Q9NPD5',\n",
       " 'Cyclin-dependent kinase inhibitor 1': 'P39689',\n",
       " 'Cellular tumor antigen p53': 'Q66HM0',\n",
       " 'Multidrug resistance-associated protein 7': 'Q5T3U5',\n",
       " 'RalA-binding protein 1': 'O55195',\n",
       " 'Canalicular multispecific organic anion transporter 2': 'O15438',\n",
       " 'DNA topoisomerase 2-alpha': 'P41516',\n",
       " 'DNA topoisomerase 2-beta': 'Q02880',\n",
       " 'Myeloperoxidase': 'P05164',\n",
       " 'Equilibrative nucleoside transporter 1': 'Q99808',\n",
       " 'Sodium/nucleoside cotransporter 1': 'O00337',\n",
       " 'Solute carrier family 28 member 3': 'Q9HAS3',\n",
       " \"5'(3')-deoxyribonucleotidase, cytosolic type\": 'Q8TCD5',\n",
       " 'Deoxycytidine kinase': 'P48769',\n",
       " 'Nucleoside diphosphate kinase A': 'P15531',\n",
       " 'Glutathione S-transferase Mu 1': 'P09488',\n",
       " 'Estrogen receptor': 'Q95131',\n",
       " 'UDP-glucuronosyltransferase 1-4': 'P22310',\n",
       " 'Dimethylaniline monooxygenase [N-oxide-forming] 3': 'P31513',\n",
       " 'Dimethylaniline monooxygenase [N-oxide-forming] 1': 'Q01740',\n",
       " 'Epidermal growth factor receptor': 'Q01279',\n",
       " 'Cetuximab': nan,\n",
       " 'Panitumumab': nan,\n",
       " 'Trastuzumab': nan,\n",
       " 'Phospholipase A2, membrane associated': 'P14423',\n",
       " 'Thiopurine S-methyltransferase': 'P51580',\n",
       " 'Ras-related C3 botulinum toxin substrate 1': 'Q6RUV5',\n",
       " 'Equilibrative nucleoside transporter 2': 'Q14542',\n",
       " 'Sodium/nucleoside cotransporter 2': 'O43868',\n",
       " 'Multidrug resistance-associated protein 5': 'O15440',\n",
       " 'Multidrug resistance-associated protein 4': 'O15439',\n",
       " 'Adenosine kinase': 'P55263',\n",
       " 'Type-1 angiotensin II receptor': 'P30556',\n",
       " 'Guanine nucleotide-binding protein G(q) subunit alpha': 'P50148',\n",
       " 'Gβ': 'P62873',\n",
       " 'Gγ': 'P59768',\n",
       " 'P2Y purinoceptor 12': 'Q9H244',\n",
       " 'Serum paraoxonase/arylesterase 1': 'P27169',\n",
       " 'Abciximab': nan,\n",
       " 'Integrin alpha-IIb': 'P08514',\n",
       " 'Integrin beta-3': 'P05106',\n",
       " 'Eptifibatide': nan,\n",
       " 'Antithrombin-III': 'P01008',\n",
       " 'Coagulation factor XII': 'Q5M879',\n",
       " 'Collagen alpha-1(I) chain': 'Q63079',\n",
       " 'Plasma kallikrein': 'P14272',\n",
       " 'Coagulation factor XI': 'Q5NTB3',\n",
       " 'Coagulation factor IX': 'F1M1M6',\n",
       " 'Coagulation factor VIII': 'A2AN88',\n",
       " 'Prothrombin': 'P18292',\n",
       " 'Coagulation factor V': 'Q28108',\n",
       " 'Coagulation factor X': 'Q63207',\n",
       " 'Fibrinogen alpha chain': 'P06399',\n",
       " 'Fibrinogen beta chain': 'Q7TME5',\n",
       " 'Fibrinogen gamma chain': 'Q68FY3',\n",
       " 'Coagulation factor XIII A chain': 'O08619',\n",
       " 'Coagulation factor XIII B chain': 'B1AY02',\n",
       " 'Plasminogen': 'Q9R0W3',\n",
       " 'Tissue-type plasminogen activator': 'P19637',\n",
       " 'Coagulation factor VII': 'Q8K3U6',\n",
       " 'Tissue factor': 'P42533',\n",
       " 'Bivalirudin': nan,\n",
       " 'Lepirudin': nan,\n",
       " 'Alteplase': nan,\n",
       " 'Anistreplase': nan,\n",
       " 'Streptokinase': nan,\n",
       " 'Tenecteplase': nan,\n",
       " 'Urokinase': nan,\n",
       " 'Reteplase': nan,\n",
       " 'Aprotinin': nan,\n",
       " 'Potassium-transporting ATPase subunit beta': 'P51164',\n",
       " 'Potassium-transporting ATPase alpha chain 1': 'P20648',\n",
       " 'Chloride intracellular channel protein 2': 'Q5M883',\n",
       " 'Carbonic anhydrase 1': 'P00915',\n",
       " 'Histamine H2 receptor': 'P25102',\n",
       " 'Gastrin': 'P04563',\n",
       " 'Gastrin/cholecystokinin type B receptor': 'P30553',\n",
       " 'Somatostatin': 'P61278',\n",
       " 'Somatostatin receptor type 4': 'P30937',\n",
       " 'Muscarinic acetylcholine receptor M3': 'P20309',\n",
       " 'Beta-1 adrenergic receptor': 'P18090',\n",
       " 'Inward rectifier potassium channel 2': 'Q64273',\n",
       " 'Inward rectifier potassium channel 4': 'O35752',\n",
       " 'G protein-activated inward rectifier potassium channel 1': 'P63251',\n",
       " 'G protein-activated inward rectifier potassium channel 4': 'P48548',\n",
       " 'Potassium voltage-gated channel subfamily D member 3': 'Q99P42',\n",
       " 'Potassium voltage-gated channel subfamily A member 5': 'Q6LEB7',\n",
       " 'Disks large homolog 1': 'Q62696',\n",
       " 'Potassium voltage-gated channel subfamily KQT member 1': 'O08655',\n",
       " 'Potassium voltage-gated channel subfamily E member 1': 'P15383',\n",
       " 'Potassium voltage-gated channel subfamily E member 2': 'Q5FVT9',\n",
       " 'Potassium voltage-gated channel subfamily H member 2': 'O08720',\n",
       " 'Putative Dol-P-Glc:Glc(2)Man(9)GlcNAc(2)-PP-Dol alpha-1,2-glucosyltransferase': 'O88788',\n",
       " 'ATP-binding cassette sub-family C member 8': 'Q09428',\n",
       " 'ATP-sensitive inward rectifier potassium channel 8': 'Q9JM50',\n",
       " 'ATP-sensitive inward rectifier potassium channel 11': 'Q9JM48',\n",
       " 'ATP-binding cassette sub-family C member 9': 'O89115',\n",
       " 'Potassium channel subfamily K member 1': 'Q9Z2T2',\n",
       " 'Sodium/calcium exchanger 1': 'P32418',\n",
       " 'Potassium/sodium hyperpolarization-activated cyclic nucleotide-gated channel 4': 'Q9QZW4',\n",
       " 'Sodium channel protein type 5 subunit alpha': 'Q14524',\n",
       " 'Beta-2-syntrophin': 'Q13425',\n",
       " 'Beta-1-syntrophin': 'Q13884',\n",
       " 'Alpha-1-syntrophin': 'Q13424',\n",
       " 'Voltage-dependent L-type calcium channel subunit alpha-1C': 'Q64178',\n",
       " 'Muscarinic acetylcholine receptor M2': 'Q9Z2Z1',\n",
       " 'Ryanodine receptor 2': 'D7UNT3',\n",
       " 'Sarcoplasmic/endoplasmic reticulum calcium ATPase 2': 'P11508',\n",
       " 'Cardiac phospholamban': 'P61016',\n",
       " 'cAMP-dependent protein kinase catalytic subunit alpha': 'Q6UA68',\n",
       " 'cAMP-dependent protein kinase type I-alpha regulatory subunit': 'P09456',\n",
       " 'Tropomyosin alpha-1 chain': 'Q63609',\n",
       " 'Tropomyosin beta chain': 'P06395',\n",
       " 'Voltage-dependent T-type calcium channel subunit alpha-1H': 'Q9EQ60',\n",
       " 'Voltage-dependent T-type calcium channel subunit alpha-1G': 'Q548R1',\n",
       " 'Epoxide hydrolase 1': 'P07099',\n",
       " 'UDP-glucuronosyltransferase 1-6': 'P19224',\n",
       " 'UDP-glucuronosyltransferase 1-8': 'Q9HAW9',\n",
       " 'UDP-glucuronosyltransferase 2B15': 'P54855',\n",
       " 'Amine oxidase [flavin-containing] B': 'Q8BW75',\n",
       " 'Neuronal acetylcholine receptor subunit alpha-3': 'P32297',\n",
       " 'Insulin': 'P01317',\n",
       " 'Hepatitis B DNA Polymerase': 'P24024',\n",
       " 'Glucocorticoid receptor': 'P04150',\n",
       " 'Heat shock protein HSP 90-alpha': 'Q91XW0',\n",
       " 'Adenylate cyclase type 2': 'P26769',\n",
       " 'cAMP-dependent protein kinase catalytic subunit beta': 'P68182',\n",
       " 'Guanine nucleotide-binding protein G(s) subunit alpha isoforms short': 'Q9Z1R8',\n",
       " 'Guanine nucleotide-binding protein G(I)/G(S)/G(O) subunit gamma-12': 'A5PJE1',\n",
       " '5-hydroxytryptamine receptor 4': 'Q63006',\n",
       " 'Ethylene-responsive transcription factor ESR1': 'Q9SAD4',\n",
       " 'Serine/threonine-protein phosphatase PP1-alpha catalytic subunit': 'P62138',\n",
       " 'Cyclic AMP-responsive element-binding protein 1': 'P15337',\n",
       " 'Corticotropin': 'P01189',\n",
       " 'Adrenocorticotropic hormone receptor': 'P34974',\n",
       " '5-hydroxytryptamine receptor 7': 'P97936',\n",
       " '5-hydroxytryptamine receptor 6': 'P31388',\n",
       " 'Adenosine receptor A2a': 'P30543',\n",
       " 'Rap guanine nucleotide exchange factor 3': 'Q9Z1C8',\n",
       " 'Ras-related protein Rap-1A': 'P62836',\n",
       " 'Dual specificity mitogen-activated protein kinase kinase 2': 'A0JN15',\n",
       " 'Serine/threonine-protein kinase B-raf': 'Q3USE9',\n",
       " 'Mitogen-activated protein kinase 1': 'P63086',\n",
       " 'ETS domain-containing protein Elk-1': 'A4GTP4',\n",
       " 'Rap guanine nucleotide exchange factor 2': 'F1MSG6',\n",
       " 'GTPase HRas': 'Q5RJJ8',\n",
       " 'Rho guanine nucleotide exchange factor 7': 'O55043',\n",
       " 'Cell division control protein 42 homolog': 'Q71TW5',\n",
       " 'Mitogen-activated protein kinase kinase kinase 4': 'Q6PDG6',\n",
       " 'Mitogen-activated protein kinase kinase kinase 1': 'Q62925',\n",
       " 'Dual specificity mitogen-activated protein kinase kinase 7': 'Q4KSH7',\n",
       " 'Mitogen-activated protein kinase 8': 'P49185',\n",
       " 'Transcription factor AP-1': 'P17325',\n",
       " 'Serine/threonine-protein kinase PAK 1': 'Q62934',\n",
       " 'Dual specificity mitogen-activated protein kinase kinase 6': 'Q5E9X2',\n",
       " 'Mitogen-activated protein kinase 11': 'Q569F1',\n",
       " 'Cyclic AMP-dependent transcription factor ATF-2': 'Q62870',\n",
       " 'Partitioning defective 6 homolog alpha': 'Q6B4M6',\n",
       " 'Protein kinase C zeta type': 'P09217',\n",
       " 'Ribosomal protein S6 kinase alpha-1': 'Q63531',\n",
       " 'Bcl2-associated agonist of cell death': 'Q9JHX1',\n",
       " '3-phosphoinositide-dependent protein kinase 1': 'G3V9W3',\n",
       " 'RAC-alpha serine/threonine-protein kinase': 'P47196',\n",
       " 'Inhibitor of nuclear factor kappa-B kinase subunit alpha': 'A7YWD1',\n",
       " 'Inhibitor of nuclear factor kappa-B kinase subunit beta': 'Q9QY78',\n",
       " 'NF-kappa-B inhibitor alpha': 'B2RYS5',\n",
       " 'Nuclear factor NF-kappa-B p105 subunit': 'Q63369',\n",
       " 'Nuclear factor NF-kappa-B p100 subunit': 'Q9WTK5',\n",
       " 'Adenosine receptor A2b': 'P29276',\n",
       " 'Vasopressin V2 receptor': 'Q53ZC4',\n",
       " 'Follicle-stimulating hormone receptor': 'Q64183',\n",
       " 'Glycoprotein hormones alpha chain': 'P70516',\n",
       " 'Follitropin subunit beta': 'P18427',\n",
       " 'Lutropin subunit beta': 'P01230',\n",
       " 'Lutropin-choriogonadotropic hormone receptor': 'Q6LDI7',\n",
       " 'Prostaglandin D2 receptor': 'Q9R261',\n",
       " 'Prostacyclin receptor': 'P43253',\n",
       " 'Myosin light chain kinase, smooth muscle': 'Q28824',\n",
       " 'Myosin light chain 3': 'P16409',\n",
       " 'Ig epsilon chain C region': 'P01855',\n",
       " 'High affinity immunoglobulin epsilon receptor subunit alpha': 'P12840',\n",
       " 'High affinity immunoglobulin epsilon receptor subunit beta': 'P13386',\n",
       " 'High affinity immunoglobulin epsilon receptor subunit gamma': 'P20411',\n",
       " 'Tyrosine-protein kinase Lyn': 'Q63320',\n",
       " 'Tyrosine-protein kinase BTK': 'Q61365',\n",
       " 'Tyrosine-protein kinase Fyn': 'Q62844',\n",
       " 'Tyrosine-protein kinase SYK': 'Q64725',\n",
       " 'GRB2-associated-binding protein 2': 'Q9EQH1',\n",
       " '1-phosphatidylinositol 4,5-bisphosphate phosphodiesterase gamma-1': 'P10686',\n",
       " 'Lymphocyte cytosolic protein 2': 'F1MGR6',\n",
       " 'Guanine nucleotide exchange factor VAV3': 'Q9JLS6',\n",
       " 'Linker for activation of T-cells family member 1': 'O70601',\n",
       " 'Protein kinase C alpha type': 'P17252',\n",
       " '[Pyruvate dehydrogenase [lipoamide]] kinase isozyme 1, mitochondrial': 'Q15118',\n",
       " 'Dual specificity mitogen-activated protein kinase kinase 4': 'P47809',\n",
       " 'Dual specificity mitogen-activated protein kinase kinase 3': 'Q91VX1',\n",
       " 'Mitogen-activated protein kinase 14': 'Q99MG4',\n",
       " 'Growth factor receptor-bound protein 2': 'Q5BKA7',\n",
       " 'Son of sevenless homolog 1': 'Q62244',\n",
       " 'GTPase NRas': 'Q4KMB1',\n",
       " 'GTPase KRas': 'P97914',\n",
       " 'RAF proto-oncogene serine/threonine-protein kinase': 'P11345',\n",
       " 'Dual specificity mitogen-activated protein kinase kinase 1': 'Q5EBD5',\n",
       " 'Cytosolic phospholipase A2': 'P50393',\n",
       " 'Interleukin-4': 'P20096',\n",
       " 'Interleukin-13': 'P42203',\n",
       " 'Interleukin-3': 'P08700',\n",
       " 'Interleukin-5': 'Q08125',\n",
       " 'Granulocyte-macrophage colony-stimulating factor': 'P48750',\n",
       " 'Tumor necrosis factor': 'Q9JI27',\n",
       " ...}"
      ]
     },
     "metadata": {},
     "execution_count": 41
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "source": [
    "test_g = g.copy()"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "source": [
    "redundant_nodes = ['Water','Hydrogen Ion','Phosphate','Oxygen','Pyrophosphate','Carbon dioxide','Hydrogen peroxide','Ammonia','Sodium']"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "source": [
    "for i in nodes:\n",
    "    if i not in c_mets_nodes and i not in c_prtn_nodes:\n",
    "        print(i,g.degree(i))\n",
    "        test_g.remove_node(i)\n",
    "    if i in redundant_nodes:\n",
    "        test_g.remove_node(i)"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "electron-transfer flavoprotein 8\n",
      "Reduced electron-transfer flavoprotein 8\n",
      "Acetyl-CoA acyltransferase 14\n",
      "Trifunctional enzyme, mitochondrial 38\n",
      "Carnitine O-palmitoyltransferase 1 6\n",
      "SubPathwayInput 152\n",
      "ATP-binding cassette sub-family D 4\n",
      "phosphatidylinositol 4,5-bisphosphate 1\n",
      "1,2-diacyl-sn-glycerol 4\n",
      "Sodium/potassium ATPase 6\n",
      "Nicotinic Acetylcholine Receptor 3\n",
      "L type Calcium channel 1\n",
      "Sodium channel 1\n",
      "Voltage Gated Potassium Channel 1\n",
      "Glutamate receptor ionotropic 1\n",
      "SubPathwayActivator 20\n",
      "Tenase complex 1\n",
      "SubPathwayInhibition 38\n",
      "Fibrinogen 2\n",
      "Fibrin (loose) 3\n",
      "Coagulation factor XIIIa 2\n",
      "Fibrin (mesh) 4\n",
      "Fibrin degradation products 3\n",
      "Tissue factor:Coagulation factor VIIa 1\n",
      "SubPathwayActivation 8\n",
      "Precursors of Prothrombin and coagulation factors VII, IX, ad X 1\n",
      "Prothrombin and coagulation factors VII, IX and X 1\n",
      "SubPathwayOutput 125\n",
      "(S)-5-Diphosphomevalonic acid 2\n",
      "Oxoglutarate dehydrogenase complex 11\n",
      "Epoxide hydratase 2 9\n",
      "AH2 2\n",
      "A  1\n",
      "Alcohol 1\n",
      "Inward rectifier potassium channel (IK1) 1\n",
      "Acetylcholine-activated potassium channel (IKACh) 1\n",
      "Voltage-gated rapidly activating transient outward delayed rectifying potassium channel (IKto) 1\n",
      "Voltage-gated ultra rapid delayed rectifying potassium channel  1\n",
      "Voltage-gated slow delayed rectifying potassium channel (IK5) 1\n",
      "Voltage-gated rapid delayed rectifying potassium channel (IKr) 1\n",
      "ATP-sensitive inward rectifier potassium channel (IKATP) 1\n",
      "Voltage-depenent L-type calcium channel 2\n",
      "Voltage-gated sodium channel (INa) 1\n",
      "Voltage-gated T-type calcium channel (ICaT) 2\n",
      "tRNA(Cys) 1\n",
      "L-Cysteinyl-tRNA(Cys) 1\n",
      "Peroxisomal membrane protein 5\n",
      "2-oxoisovalerate dehydrogenase 7\n",
      "Dihydrolipoyl dehydrogenase E3 5\n",
      "Organic Anion 1\n",
      "Organic Cation 1\n",
      "Epithelial Sodium Channel 1\n",
      "L-Amino acid 4\n",
      "SLC3A2/4F2hc Large neutral amino acids transporter small subunit 2 1\n",
      "Large Neutral Amino Acids 2\n",
      "SLC3A2/4F2hc Large neutral amino acids transporter small subunit 1 1\n",
      "Imino acids 1\n",
      "SLC3A2/4F2hc Y+L amino acid transporter 1 2\n",
      "SLC3A2/4F2hc Y+L amino acid transporter 2 2\n",
      "Neutral and basic amino acid transport protein rBAT/B(0,+)-type amino acid transporter 1 1\n",
      "Lactose synthase 4\n",
      "Acceptor 6\n",
      "Reduced acceptor 6\n",
      "A 1\n",
      "Succinyl-CoA ligase 7\n",
      "Pyruvate dehydrogenase complex 8\n",
      "Isocitrate dehydrogenase 8\n",
      "Succinate dehydrogenase 8\n",
      "Glutamine--fructose-6-phosphate aminotransferase 6\n",
      "tRNA(Gln) 1\n",
      "L-Glutaminyl-tRNA(Gln) 1\n",
      "tRNA(Glu) 2\n",
      "L-Glutamyl-tRNA(Glu) 2\n",
      "branched-chain alpha-keto dehydrogenase complex 6\n",
      "Propionyl-CoA carboxylase 6\n",
      "Biotin-Carboxyl Carrying Protein 2\n",
      "tRNA(Pro) 1\n",
      "L-Prolyl-tRNA(Pro) 1\n",
      "Prolyl 4-hydroxylase 6\n",
      "tRNA(Arg) 2\n",
      "L-Arginyl-tRNA(Arg) 2\n",
      "Nitric-oxide synthase, brain 7\n",
      "SubPathwayInhibitor 22\n",
      " Ribose-phosphate pyrophosphokinase 3 6\n",
      "Ribonucleoside-diphosphate reductase 1 6\n",
      "Deoxyguanosine kinase 4\n",
      "Guanylate cyclase 3\n",
      "Ferricytochrome c 1\n",
      "Ferrocytochrome c 1\n",
      "Pyruvate dehydrogenase E1 5\n",
      "Phenylalanine--tRNA ligase 7\n",
      "tRNA(Phe) 1\n",
      "L-Phenylalanyl-tRNA(Phe) 1\n",
      "tRNA(Tyr) 1\n",
      "L-Tyrosyl-tRNA(Tyr) 1\n",
      "H(+)/K(+) ATPase subunit 6\n",
      "L-Seryl-tRNA 1\n",
      "L-Seryl-tRNA(Ser) 1\n",
      "Glycyl-tRNA 1\n",
      "Glycyl-tRNA(Gly) 1\n",
      "Reduced adrenal ferredoxin 3\n",
      "Oxidized adrenal ferredoxin 3\n",
      "Soluble guanylyl cyclase 2\n",
      "Potassium voltage-gated channel, KQT-like subfamily 1\n",
      "tRNA(His) 2\n",
      "L-Histidyl-tRNA(His) 2\n",
      "tRNA(Asn) 1\n",
      "L-Asparaginyl-tRNA(Asn) 1\n",
      "a reduced glutaredoxin 1\n",
      "an oxidized glutaredoxin 1\n",
      "a purine ribonucleoside 1\n",
      "a purine base 1\n",
      "reduced electron acceptor 2\n",
      "oxidized electron acceptor 2\n",
      "PIP2 4\n",
      "PIP3 1\n",
      "tRNA(Ala) 2\n",
      "L-Alanyl-tRNA(Ala) 2\n",
      "long-chain fatty acyl-CoA  4\n",
      "1-acyl-sn-glycerol 3-phosphate  3\n",
      "1,2-diacyl-sn-glycerol 3-phosphate 4\n",
      "triglyceride 1\n",
      "Bifunctional UDP-N-acetyl glucosamine 2-epimerase / N-acetylmannosamine kinase 7\n",
      "Reduced flavoprotein 2\n",
      "Oxidized flavoprotein 2\n",
      "Carboxylate 3\n",
      "Thiamin pyrophosphokinase 6\n",
      "tRNA(Val) 1\n",
      "L-Valyl-tRNA(Val) 1\n",
      "tRNA(Trp) 2\n",
      "L-Tryptophyl-tRNA(Trp) 2\n",
      "oxidized thioredoxin  2\n",
      "reduced thioredoxin 2\n",
      "tRNA(Met) 1\n",
      "L-Methionyl-tRNA(Met) 2\n",
      "Phosphatidylcholine 4\n",
      "tRNA(Lys) 1\n",
      "L-Lysyl-tRNA(Lys) 1\n",
      "quinol 1\n",
      "thioredoxin disulfide 1\n",
      "thioredoxin 1\n",
      "Thioredoxin disulfide 1\n",
      "Guanine nucleotide-binding protein alpha-beta-gamma 13\n",
      "Guanine nucleotide-binding protein gamma-beta 1\n",
      "3-oxoacid CoA-transferase 1 6\n",
      "Peptidyl-4-carboxyglutamate 1\n",
      "Peptidyl-glutamate 1\n",
      "DG 2\n",
      "IP3 3\n",
      "Phosphatidylinositol 4,5-bisphosphate 3-kinase 2\n",
      "Adenylate cyclase bound to G(s) subunit alpha 3\n",
      "Methotrexate polyglutamate 1\n",
      "N-formylmethionyl-tRNA(fMet) 1\n",
      "glutamate-pyruvate aminotransferase 4\n",
      "mutant succinate dehydrogenase 6\n",
      "Hydrogen ion 1\n",
      "Glutamine transporter, mitochondrial 1\n",
      "Fatty Acid 1\n",
      "Cytochrome P450 2\n",
      "GTP cyclohydrolase 8\n",
      "R 1\n",
      "CH3-R 1\n",
      "Selenomethionyl-tRNA(Met) 1\n",
      "Nitric-oxide synthase, endothelial 4\n",
      "tRNA(Leu) 1\n",
      "L-Leucyl-tRNA(Leu) 1\n",
      "tRNA(Thr) 1\n",
      "L-Threonyl-tRNA(Thr) 1\n",
      "CDP-Ethanolamine 2\n",
      "an L-1-phosphatidylserine 2\n",
      "an L-1-phosphatidylethanolamine 3\n",
      "tRNA(Asp) 1\n",
      "L-Aspartyl-tRNA(Asp) 1\n",
      "A (5-L-glutamyl)-peptide 1\n",
      "Peptide 1\n",
      "a CDP-diacylglycerol  3\n",
      "an L-1-phosphatidyl-glycerol  2\n",
      "a cardiolipin 1\n",
      "an L-1-phosphatidylglycerol-phosphate  2\n",
      " solute carrier family 44 1\n",
      "Inhibitor of nuclear factor kappa-B kinase 2\n",
      "Dihydroxyacetone phosphate acyltransferase and Alkyldihydroxyacetonephosphate synthase 7\n",
      "Activated Phosphatidylinositol 4,5-bisphosphate 3-kinase complex 2\n",
      "GRB2-SOS-IRS-SHC complex 3\n",
      "RAF proto-oncogene serine/threonine-protein kinase recruited by GTPase HRas 2\n",
      "Bile Acid  3\n",
      "Phosphatidyl-N-dimethylethanolamine 1\n",
      "prolyl hydroxylase 6\n",
      "tRNA(Ile) 1\n",
      "L-Isoleucyl-tRNA(Ile) 1\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "source": [
    "attibutes_type_dict = {}"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "source": [
    "for node in test_g.nodes():\n",
    "    if node in c_mets_nodes:\n",
    "        attibutes_type_dict[node]='metabolite'\n",
    "    elif node in c_prtn_nodes:\n",
    "        attibutes_type_dict[node]='protein'"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "source": [
    "nx.set_node_attributes(test_g, attibutes_type_dict, name=\"type\")"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "source": [
    "nx.write_gexf(test_g, \"test_g.gexf\")"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "source": [
    "def plot_degree_dist(G,nodes):\n",
    "    degrees = [G.degree(n) for n in nodes]\n",
    "    plt.hist(degrees)\n",
    "    plt.show()\n",
    "\n",
    "met_degrees = [(n,g.degree(n)) for n in c_mets_nodes]\n",
    "prtn_degrees = [(n,g.degree(n)) for n in c_prtn_nodes]"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "source": [
    "met_degrees_counts = [i[1] for i in met_degrees]\n",
    "prtn_degrees_counts = [i[1] for i in prtn_degrees]"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "source": [
    "np.median(met_degrees_counts)"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "2.0"
      ]
     },
     "metadata": {},
     "execution_count": 163
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "source": [
    "np.median(prtn_degrees_counts)"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "5.0"
      ]
     },
     "metadata": {},
     "execution_count": 164
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "source": [
    "met_degrees_sort = sorted(met_degrees, key = lambda x: x[1],reverse=True)\n",
    "prtn_degrees_sort = sorted(prtn_degrees, key = lambda x: x[1],reverse=True)"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "source": [
    "met_degrees_sort[int(len(met_degrees_sort)/2)]"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "('D-Maltose', 2)"
      ]
     },
     "metadata": {},
     "execution_count": 166
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "source": [
    "prtn_degrees_sort[int(len(prtn_degrees_sort)/2)]"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "('D-3-phosphoglycerate dehydrogenase', 5)"
      ]
     },
     "metadata": {},
     "execution_count": 167
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "source": [
    "c_met_degrees = [0]*11\n",
    "c_prtn_degrees = [0]*11\n",
    "for c,i in met_degrees:\n",
    "    if i <= 10:\n",
    "        c_met_degrees[i-1] += 1\n",
    "    else:\n",
    "        c_met_degrees[-1] += 1\n",
    "for c,i in prtn_degrees:\n",
    "    if i <= 10:\n",
    "        c_prtn_degrees[i-1] += 1\n",
    "    else:\n",
    "        c_prtn_degrees[-1] += 1"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "source": [
    "c_met_degrees"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "[387, 367, 170, 109, 65, 51, 26, 14, 9, 8, 55]"
      ]
     },
     "metadata": {},
     "execution_count": 169
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "source": [
    "c_prtn_degrees"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "[57, 92, 78, 173, 69, 112, 75, 49, 28, 17, 105]"
      ]
     },
     "metadata": {},
     "execution_count": 170
    }
   ],
   "metadata": {}
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "3fe952b0ba1daa6ee3393e62a893871d88149d9d44cf2ee1c1e3a6730bdcafed"
  },
  "kernelspec": {
   "name": "python3",
   "display_name": "Python 3.8.8 64-bit ('base': conda)"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "317.2146911621094px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}